<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Use case examples" href="../usecases.html" /><link rel="prev" title="Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities" href="TCR_motifs.html" />
        <link rel="canonical" href="https://uio-bmi.github.io/ligo/tutorials/epitope-specific_TCR_AIRRs.html" />

    <link rel="shortcut icon" href="../_static/favicon.ico"/><!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>Simulating epitope-specific TCRs and validating their properties - LIgO 1.0.11 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/immuneml.css?v=3985a9e7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">LIgO 1.0.11 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">LIgO 1.0.11 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 has-children"><a class="reference internal" href="../quickstart.html">Quickstart</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Quickstart</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/receptor_level_simulation.html">How to use LIgO for receptor-level simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/repertoire_level_simulation.html">How to use LIgO for repertoire-level simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/overview_of_ligo_parameters.html">Overview of LIgO simulation parameters</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../installation.html">Installing LIgO</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Installing LIgO</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="how_to_define_immune_signals_and_events.html">How to define immune signals and immune events</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_choose_simulation_strategy.html">How to choose between rejection sampling and signal implantation</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_check_feasibility_of_sim_params.html">How to check feasibility of the simulation parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_handle_the_maximum_iterations_were_reached_error.html">How to handle the maximum iterations were reached error</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulation_with_custom_signal_functions.html">Simulation with custom signal functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_simulate_paired_chain_data.html">Paired chain simulations in LIgO</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_simulate_co-occuring_signals.html">How to simulate co-occuring immune signals</a></li>
<li class="toctree-l2"><a class="reference internal" href="TCR_motifs.html">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Simulating epitope-specific TCRs and validating their properties</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../usecases.html">Use case examples</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Use case examples</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../usecases/usecase1.html">Manuscript use case 1: Out-of-distribution receptor-level simulation using LIgO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/usecase2.html">Manuscript use case 2: Limitations of conventional encoding schemes for repertoire-level binary classification when immune signals co-occur within the same AIR</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../specification.html">YAML specification</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of YAML specification</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/tutorials/epitope-specific_TCR_AIRRs.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="simulating-epitope-specific-tcrs-and-validating-their-properties">
<h1>Simulating epitope-specific TCRs and validating their properties<a class="headerlink" href="#simulating-epitope-specific-tcrs-and-validating-their-properties" title="Link to this heading">¶</a></h1>
<p>During the development of a new epitope-TCR prediction method, it might be useful to simulate training data to test it or study potential biases in model performance. In this tutorial, we simulate epitope-specific TCR data, i.e., a collection of TCRs derived from different individuals that are all recognizing the same epitope.</p>
<p>In this tutorial, epitope-specific TCR beta receptors are simulated from three user-defined motifs, which are considered to describe the subsequences within TCRs that are epitope-specific. Next, we ensure that the simulated data is biologically relevant by comparing the simulated TCRs to the experimental epitope-specific data. This use case is inspired by the manucript <strong>“Revealing the hidden sequence distribution of epitope-specific TCR repertoires and its influence on machine learning model performance”</strong>, see the manuscript text (<a class="reference external" href="https://www.biorxiv.org/content/10.1101/2024.10.21.619364v1">biorxiv</a>) for more details.</p>
<section id="simulating-of-the-epitope-specific-tcrs">
<h2>Simulating of the epitope-specific TCRs<a class="headerlink" href="#simulating-of-the-epitope-specific-tcrs" title="Link to this heading">¶</a></h2>
<p>We defined an epitope-specific motif using the long seeds created in the previous tutorial <a class="reference internal" href="TCR_motifs.html#constructing-ligo-motifs-inspired-by-a-database-of-tcr-sequences-with-known-antigen-specificities"><span class="std std-ref">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</span></a> (other details on transforming known epitope-specific TCRs from VDJdb into LIgO motifs can be found in the same tutorial). The rejection sampling method was used to avoid introducing additional artifacts to LIgO-simulated TCRs.</p>
<p>The yaml file below describes simulation parameters used for LIgO simulation. You can find more details on how the parameters for the simulation were chosen in the tutorial <a class="reference internal" href="TCR_motifs.html#constructing-ligo-motifs-inspired-by-a-database-of-tcr-sequences-with-known-antigen-specificities"><span class="std std-ref">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</span></a> and find more information about how to run receptor-level simulation using the quickstart <a class="reference internal" href="../quickstart/receptor_level_simulation.html#how-to-use-ligo-for-receptor-level-simulation"><span class="std std-ref">How to use LIgO for receptor-level simulation</span></a>.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">definitions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">motifs</span><span class="p">:</span><span class="w"> </span><span class="c1"># define motifs based on long seeds</span>
<span class="w">    </span><span class="nt">motif1</span><span class="p">:</span>
<span class="w">      </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span>
<span class="w">        </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.1</span>
<span class="w">        </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span>
<span class="w">        </span><span class="nt">2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span>
<span class="w">      </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ELSGINQP</span>
<span class="w">    </span><span class="nt">motif2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span>
<span class="w">        </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.1</span>
<span class="w">        </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span>
<span class="w">        </span><span class="nt">2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span>
<span class="w">      </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">SPAGGTYE</span>
<span class="w">    </span><span class="nt">motif3</span><span class="p">:</span>
<span class="w">      </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span>
<span class="w">        </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.1</span>
<span class="w">        </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span>
<span class="w">        </span><span class="nt">2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span>
<span class="w">      </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">SGGDVREE</span>
<span class="w">  </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">    </span><span class="nt">signal1</span><span class="p">:</span>
<span class="w">      </span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">motif1</span>
<span class="w">      </span><span class="nt">sequence_position_weights</span><span class="p">:</span>
<span class="w">        </span><span class="s">&#39;104&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span><span class="w"> </span><span class="c1"># we did not want to start the motif at the first position, i.e. IMGT position 104</span>
<span class="w">    </span><span class="nt">signal2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">motif2</span>
<span class="w">      </span><span class="nt">sequence_position_weights</span><span class="p">:</span>
<span class="w">        </span><span class="s">&#39;104&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span><span class="w"> </span><span class="c1"># we did not want to start the motif at the first position, i.e. IMGT position 104</span>
<span class="w">    </span><span class="nt">signal3</span><span class="p">:</span>
<span class="w">      </span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">motif3</span>
<span class="w">      </span><span class="nt">sequence_position_weights</span><span class="p">:</span>
<span class="w">        </span><span class="s">&#39;104&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span><span class="w"> </span><span class="c1"># we did not want to start the motif at the first position, i.e. IMGT position 104</span>
<span class="w">  </span><span class="nt">simulations</span><span class="p">:</span>
<span class="w">    </span><span class="nt">sim1</span><span class="p">:</span>
<span class="w">      </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">      </span><span class="nt">paired</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">      </span><span class="nt">sequence_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">amino_acid</span>
<span class="w">      </span><span class="nt">simulation_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">RejectionSampling</span>
<span class="w">      </span><span class="nt">sim_items</span><span class="p">:</span>
<span class="w">        </span><span class="nt">var1</span><span class="p">:</span>
<span class="w">          </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">            </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">            </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">          </span><span class="nt">is_noise</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">          </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">300</span><span class="w"> </span><span class="c1"># simulate 300 TCRs</span>
<span class="w">          </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">            </span><span class="nt">signal1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"> </span><span class="c1"># all TCRs having signal1</span>
<span class="w">        </span><span class="nt">var2</span><span class="p">:</span>
<span class="w">          </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">            </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">            </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">          </span><span class="nt">is_noise</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">          </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">300</span><span class="w"> </span><span class="c1"># simulate 300 TCRs</span>
<span class="w">          </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">            </span><span class="nt">signal2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"> </span><span class="c1"># all TCRs having signal2</span>
<span class="w">        </span><span class="nt">var3</span><span class="p">:</span>
<span class="w">          </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">            </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">            </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">          </span><span class="nt">is_noise</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">          </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">300</span><span class="w"> </span><span class="c1"># simulate 300 TCRs</span>
<span class="w">          </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">            </span><span class="nt">signal3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"> </span><span class="c1"># all TCRs having signal3</span>
<span class="nt">instructions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">inst1</span><span class="p">:</span>
<span class="w">    </span><span class="nt">export_p_gens</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">    </span><span class="nt">max_iterations</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2000</span>
<span class="w">    </span><span class="nt">number_of_processes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">8</span>
<span class="w">    </span><span class="nt">sequence_batch_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10000</span>
<span class="w">    </span><span class="nt">simulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sim1</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">LigoSim</span>
<span class="nt">output</span><span class="p">:</span>
<span class="w">  </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">HTML</span>
</pre></div>
</div>
</section>
<section id="inspecting-the-simulated-tcrs">
<h2>Inspecting the simulated TCRs<a class="headerlink" href="#inspecting-the-simulated-tcrs" title="Link to this heading">¶</a></h2>
<p>LIgO-simulated data can only resemble the general structure of epitope-specific TCRs. The actual specificities of these TCRs cannot be simulated, and thus the final simulated data are not guaranteed to contain only TCRs recognizing the same epitope. Otherwise, the problem of predicting epitope-TCR binding would be solved, and no models should be trained anymore. However, several methods can be used to validate LIgO-simulated data and compare it to the experimental data:</p>
<section id="method-1-validating-that-ligo-simulated-clusters-accurately-represent-immune-signals">
<h3>Method 1: Validating that LIgO-simulated clusters accurately represent immune signals<a class="headerlink" href="#method-1-validating-that-ligo-simulated-clusters-accurately-represent-immune-signals" title="Link to this heading">¶</a></h3>
<p>While LIgO ensures that each signal-specific receptor includes an immune signal, the distribution of these signals in the final repertoire may occasionally diverge from the initial definition. For example, if a signal k-mer allows for a large Hamming distance, it could introduce high variations in the data and thus the initial k-mer signal would be lost.</p>
<p>To address this, we recommend verifying that the motifs in the final dataset accurately reflect the initial LIgO motifs.</p>
<p><strong>How to verify motif in LIgO-simulated data:</strong></p>
<ol class="arabic simple">
<li><p>Cluster the simulated data. Any clustering method can be utilized, in this tutorial we demonstrate clustering using the clusTCR tool (Valkiers et al., 2021). See the script below for more details</p></li>
<li><p>Compare cluster motifs to initial LIgO motifs. This comparison will reveal if the final repertoire accurately captures the initial immune signals.</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">clustcr</span> <span class="kn">import</span> <span class="n">Clustering</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Read LIgO-simulated data</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;./sim/inst1/exported_dataset/airr/batch1.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># Cluster training data for selected epitope using clusTCR</span>
<span class="n">clustering</span> <span class="o">=</span> <span class="n">Clustering</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;MCL&#39;</span><span class="p">)</span>
<span class="n">clustered_data</span> <span class="o">=</span> <span class="n">clustering</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;junction_aa&#39;</span><span class="p">])</span>

<span class="c1"># Save clusTCR motifs sorted by cluster size</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">clustered_data</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">motifs</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;size&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">motifs</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;clustcr_motifs.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>The clustcr_motifs.csv file will contain motifs saved in clusTCR format. You can read more about clusTCR motif format in the <a class="reference external" href="https://svalkiers.github.io/clusTCR/">clusTCR documentation</a>. Briefly, the motif uses upper-case for highly conserved amino acids (frequency &gt; 0.7) and lower-case for moderately conserved ones. If two amino acids are equally frequent, they’re in brackets ([ ]), and less significant positions use a dot (.) as a wildcard.</p>
<p>The motifs for the 10 largest clusters are given in the table 1 below. We can see a clear overlap between the clusTCR motifs and the original seeds. Thus, the original motif has been successfully simulated within the entire repertoire using long seeds.</p>
<div class="table-wrapper docutils container" id="id3">
<table class="docutils align-default" id="id3">
<caption><span class="caption-text">Table 1: ClusTCR motifs of top-10 largest clusters when TCR dataset is simulated using long seeds</span><a class="headerlink" href="#id3" title="Link to this table">¶</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>clusTCR motif</p></th>
<th class="head"><p>LIgO seed</p></th>
<th class="head"><p>cluster size</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CASSp.GGtYEQYF</p></td>
<td><p>SPAGGTYE</p></td>
<td><p>59</p></td>
</tr>
<tr class="row-odd"><td><p>CAS[SR].gGTYEQYF</p></td>
<td><p>SPAGGTYE</p></td>
<td><p>19</p></td>
</tr>
<tr class="row-even"><td><p>CASSLSG.NQPQHF</p></td>
<td><p>ELSGINQP</p></td>
<td><p>16</p></td>
</tr>
<tr class="row-odd"><td><p>CASSL.GINQPQHF</p></td>
<td><p>ELSGINQP</p></td>
<td><p>9</p></td>
</tr>
<tr class="row-even"><td><p>CASSAGG.YEQYF</p></td>
<td><p>SPAGGTYE</p></td>
<td><p>7</p></td>
</tr>
<tr class="row-odd"><td><p>CASSGG.VRYEQYF</p></td>
<td><p>SGGDVREE</p></td>
<td><p>6</p></td>
</tr>
<tr class="row-even"><td><p>CASP[GP]GG.YEQYF</p></td>
<td><p>SPAGGTYE</p></td>
<td><p>6</p></td>
</tr>
<tr class="row-odd"><td><p>CASSE.SGSNQPQHF</p></td>
<td><p>ELSGINQP</p></td>
<td><p>5</p></td>
</tr>
<tr class="row-even"><td><p>CASSPGtGTYEQYF</p></td>
<td><p>SPAGGTYE</p></td>
<td><p>4</p></td>
</tr>
<tr class="row-odd"><td><p>CASSvAGGTGELFF</p></td>
<td><p>SPAGGTYE</p></td>
<td><p>4</p></td>
</tr>
</tbody>
</table>
</div>
<p>In contrast, if we simulate TCR data using short seeds (see <a class="reference internal" href="TCR_motifs.html#constructing-ligo-motifs-inspired-by-a-database-of-tcr-sequences-with-known-antigen-specificities"><span class="std std-ref">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</span></a> for long and short seeds defenition), we observe that the seeds are not well represented within the clusTCR motifs (see Table 2 below). Clustering might be a result of other common similarities outside of the predefined motif. We thus lost track of the original motif after simulation.</p>
<div class="table-wrapper docutils container" id="id4">
<table class="docutils align-default" id="id4">
<caption><span class="caption-text">Table 2: ClusTCR motifs of top-10 largest clusters when TCR dataset is simulated using long seeds
 :header-rows: 1</span><a class="headerlink" href="#id4" title="Link to this table">¶</a></caption>
<tbody>
<tr class="row-odd"><td><p>clusTCR motif</p></td>
<td><p>LIgO seed</p></td>
<td><p>cluster size</p></td>
</tr>
<tr class="row-even"><td><p>CA..YEQYF</p></td>
<td><p>PAG,DVR,SGI</p></td>
<td><p>15</p></td>
</tr>
<tr class="row-odd"><td><p>CAs.yEQYF</p></td>
<td><p>PAG,DVR,SGI</p></td>
<td><p>14</p></td>
</tr>
<tr class="row-even"><td><p>CA.T[AP]YEQYF</p></td>
<td><p>PAG,DVR</p></td>
<td><p>9</p></td>
</tr>
<tr class="row-odd"><td><p>CArDEQYF</p></td>
<td><p>DVR</p></td>
<td><p>8</p></td>
</tr>
<tr class="row-even"><td><p>CAS..ETQYF</p></td>
<td><p>SGI</p></td>
<td><p>8</p></td>
</tr>
<tr class="row-odd"><td><p>CAS.tYEQYF</p></td>
<td><p>SGI</p></td>
<td><p>6</p></td>
</tr>
<tr class="row-even"><td><p>C[RT]DYEQYF</p></td>
<td><p>DVR</p></td>
<td><p>5</p></td>
</tr>
<tr class="row-odd"><td><p>CA[KT][SR]ETQYF</p></td>
<td><p>DVR,SGI</p></td>
<td><p>5</p></td>
</tr>
<tr class="row-even"><td><p>C[GV]G[QL]YEQYF</p></td>
<td><p>SGI</p></td>
<td><p>5</p></td>
</tr>
<tr class="row-odd"><td><p>CAR.TDTQYF</p></td>
<td><p>DVR</p></td>
<td><p>5</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="method-2-visual-investigation-using-umap">
<h3>Method 2: Visual investigation using UMAP<a class="headerlink" href="#method-2-visual-investigation-using-umap" title="Link to this heading">¶</a></h3>
<p>As shown by Gielis et al (<a class="reference external" href="https://www.biorxiv.org/content/10.1101/2024.10.21.619364v1">biorxiv</a>), epitope-specific TCRs are spread out over TCR sequence space (figure 2 of the manuscript “Revealing the hidden sequence distribution of epitope-specific TCR repertoires and its influence on machine learning model performance”). This observation can be useful for visual examination whether LIgO-simulated data shows natural groupings among TCRs derived from distinct LIgO seeds.</p>
<p>Figure below shows a UMAP plot constructed from a LIgO long-seed data simulation. Notably, the simulated TCRs are primarily clustered by their originating seed, which may not fully align with patterns observed in biological data. By contrast, Figure 4 in the Gielis et al. manuscript illustrates a UMAP plot where TCRs simulated from various seeds are interspersed, more closely resembling the complex distribution seen in real TCR data.</p>
<p>The UMAP was generated using RapTCR tool (Van Deuren et al., 2023). For more information on UMAP construction, please refer to the <a class="reference external" href="https://vincentvandeuren.github.io/RapTCR_docs/visualizing/UMAP.html">RapTCR documentation</a> and the Gielis et al. manuscript (<a class="reference external" href="https://www.biorxiv.org/content/10.1101/2024.10.21.619364v1">biorxiv</a>).</p>
<p>Figure 1: UMAP plot of LIgO-simulated data using long seeds. TCRs corresponding to different immune signals are shown in different colors.</p>
<a class="reference internal image-reference" href="../_images/UMAP.png"><img alt="../_images/UMAP.png" src="../_images/UMAP.png" style="width: 1500px;" />
</a>
</section>
<section id="method-3-comparing-descriptive-statistics-with-experimental-data">
<h3>Method 3: Comparing descriptive statistics with experimental data<a class="headerlink" href="#method-3-comparing-descriptive-statistics-with-experimental-data" title="Link to this heading">¶</a></h3>
<p>Due to the complex, high-dimensional nature of immune repertoires, direct comparisons between simulated and experimental datasets are challenging. However, descriptive statistics provide a practical approach for assessing how closely the simulated data aligns with experimental data. The basic statistics to consider is the is the CDR3 length distribution. As shown on the figure below, the CDR3s simulated with long seeds are more closely match the CDR3 length of experimental TCRs used to construct those seeds (14AA). By contrast, simulations based on shorter seeds produce CDR3s that are shorter than those in the original dataset. Users can further refine CDR3 length distributions using the sequence_len_limits parameter in the simulation configuration.</p>
<a class="reference internal image-reference" href="../_images/length_distr.png"><img alt="../_images/length_distr.png" src="../_images/length_distr.png" style="width: 1500px;" />
</a>
<p>When LIgO signals are constructed from a larger initial TCR dataset, as demonstrated in the <span class="xref std std-ref">Enhanced approach: defining motif based on PWM</span> or in the Gielis et al. manuscript, users can incorporate additional descriptive statistics to compare simulated and experimental data, such as:</p>
<ul class="simple">
<li><p>Basic repertoire statistics: sequence diversity (e.g., Gini index) and or V/J gene usage</p></li>
<li><p>Clustering-based statistics: distribution of cluster sizes within the repertoire</p></li>
</ul>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../usecases.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Use case examples</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="TCR_motifs.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Simulating epitope-specific TCRs and validating their properties</a><ul>
<li><a class="reference internal" href="#simulating-of-the-epitope-specific-tcrs">Simulating of the epitope-specific TCRs</a></li>
<li><a class="reference internal" href="#inspecting-the-simulated-tcrs">Inspecting the simulated TCRs</a><ul>
<li><a class="reference internal" href="#method-1-validating-that-ligo-simulated-clusters-accurately-represent-immune-signals">Method 1: Validating that LIgO-simulated clusters accurately represent immune signals</a></li>
<li><a class="reference internal" href="#method-2-visual-investigation-using-umap">Method 2: Visual investigation using UMAP</a></li>
<li><a class="reference internal" href="#method-3-comparing-descriptive-statistics-with-experimental-data">Method 3: Comparing descriptive statistics with experimental data</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=9306f2a9"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>