<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="How to choose between rejection sampling and signal implantation" href="how_to_choose_simulation_strategy.html" /><link rel="prev" title="Tutorials" href="../tutorials.html" />
        <link rel="canonical" href="https://uio-bmi.github.io/ligo/tutorials/how_to_define_immune_signals_and_events.html" />

    <link rel="shortcut icon" href="../_static/favicon.ico"/><!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>How to define immune signals and immune events - LIgO 1.0.11 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/immuneml.css?v=3985a9e7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">LIgO 1.0.11 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">LIgO 1.0.11 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 has-children"><a class="reference internal" href="../quickstart.html">Quickstart</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Quickstart</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/receptor_level_simulation.html">How to use LIgO for receptor-level simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/repertoire_level_simulation.html">How to use LIgO for repertoire-level simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/overview_of_ligo_parameters.html">Overview of LIgO simulation parameters</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../installation.html">Installing LIgO</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Installing LIgO</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">How to define immune signals and immune events</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_choose_simulation_strategy.html">How to choose between rejection sampling and signal implantation</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_check_feasibility_of_sim_params.html">How to check feasibility of the simulation parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_handle_the_maximum_iterations_were_reached_error.html">How to handle the maximum iterations were reached error</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulation_with_custom_signal_functions.html">Simulation with custom signal functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_simulate_paired_chain_data.html">Paired chain simulations in LIgO</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_simulate_co-occuring_signals.html">How to simulate co-occuring immune signals</a></li>
<li class="toctree-l2"><a class="reference internal" href="TCR_motifs.html">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</a></li>
<li class="toctree-l2"><a class="reference internal" href="epitope-specific_TCR_AIRRs.html">Simulating epitope-specific TCRs and validating their properties</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../usecases.html">Use case examples</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Use case examples</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../usecases/usecase1.html">Manuscript use case 1: Out-of-distribution receptor-level simulation using LIgO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/usecase2.html">Manuscript use case 2: Limitations of conventional encoding schemes for repertoire-level binary classification when immune signals co-occur within the same AIR</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../specification.html">YAML specification</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of YAML specification</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/tutorials/how_to_define_immune_signals_and_events.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="how-to-define-immune-signals-and-immune-events">
<h1>How to define immune signals and immune events<a class="headerlink" href="#how-to-define-immune-signals-and-immune-events" title="Link to this heading">¶</a></h1>
<p>Adaptive immune receptors (AIRs) specifically recognize various antigens, so they can be labeled with antigen specificity.
Adaptive immune receptor repertoires (AIRRs) coming from individual patients can in a similar way be labeled with individual’s disease state(s).
To facilitate development and benchmarking of AIRR-related machine learning models, we focus on biological <strong>immune events</strong> (e.g., disease, allergy, vaccination)
and <strong>immune signals</strong> that reflect the binding rules to the immune event antigens. Because of this formalization,
multiple immune signals can be associated with a single immune event.</p>
<p>In this tutorial, we describe this formalization we make in more details and connect it to the simulation specification.
More details on each of the options in the specification can be found under <a class="reference internal" href="../specification.html#yaml-specification"><span class="std std-ref">YAML specification</span></a> page.</p>
<p>To define an immune signal on the AIR level, we define the following:</p>
<ul class="simple">
<li><p>a set of <strong>motifs</strong> determining the content of the receptor sequence, where motifs are defined as a distribution over amino acids or nucleotides,</p></li>
<li><p>motif locations in the CDR3,</p></li>
<li><p>V gene,</p></li>
<li><p>J gene.</p></li>
</ul>
<section id="motifs">
<h2>Motifs<a class="headerlink" href="#motifs" title="Link to this heading">¶</a></h2>
<p>LIgO allows for two types of motifs:</p>
<ul class="simple">
<li><p>a motif based on the seed string with possible gaps and allowed variations from the seed and</p></li>
<li><p>a positional weight matrix describing multinomial distribution of amino acids or nucleotides over the motif positions.</p></li>
</ul>
<section id="seed-motif">
<h3>Seed motif<a class="headerlink" href="#seed-motif" title="Link to this heading">¶</a></h3>
<p>Here is an example of the motif definition based on a seed. It is possible to define seed that can optionally contain a
gap denoted with <cite>/</cite> sign, minimum and maximum size of the gap, probabilities of different Hamming distances (how many
letters in the motif can be changed and with what probability), position weights (probabilities that a letter in the
seed will be changed for each letter), and alphabet weights (which letters to pick for replacement to implement the
required Hamming distance).</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_simple_motif</span><span class="p">:</span><span class="w"> </span><span class="c1"># the name of the motif used for reference later</span>
<span class="w">      </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AAA</span><span class="w"> </span><span class="c1"># motif is always AAA</span>
<span class="nt">my_gapped_motif</span><span class="p">:</span><span class="w"> </span><span class="c1"># the name of the more complex motif where / sign denotes a possible gap location</span>
<span class="w">      </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AA/A</span><span class="w"> </span><span class="c1"># this motif can be AAA, AA_A, CAA, CA_A, DAA, DA_A, EAA, EA_A</span>
<span class="w">      </span><span class="nt">min_gap</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span><span class="w"> </span><span class="c1"># how many gaps can there be: min 0 and max 1</span>
<span class="w">      </span><span class="nt">max_gap</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">      </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span><span class="w"> </span><span class="c1"># it can have a max of 1 substitution</span>
<span class="w">          </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span>
<span class="w">          </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.3</span>
<span class="w">      </span><span class="nt">position_weights</span><span class="p">:</span><span class="w"> </span><span class="c1"># note that index 2, the position of the gap, is excluded from position_weights</span>
<span class="w">          </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"> </span><span class="c1"># only first position can be changed</span>
<span class="w">          </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">          </span><span class="nt">3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">      </span><span class="nt">alphabet_weights</span><span class="p">:</span><span class="w"> </span><span class="c1"># the first A can be replaced by C, D or E</span>
<span class="w">          </span><span class="nt">C</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>
<span class="w">          </span><span class="nt">D</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>
<span class="w">          </span><span class="nt">E</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span>
</pre></div>
</div>
</section>
<section id="positional-weight-matrix">
<h3>Positional weight matrix<a class="headerlink" href="#positional-weight-matrix" title="Link to this heading">¶</a></h3>
<p>Motifs can alternatively be defined as positional weight matrices (PWMs). For importing PWMs (and later annotation of
sequences with motifs defined as PWMs), LIgO relies on bionumpy library and supports the formats supported by the library.
For more information, see <a class="reference external" href="https://bionumpy.github.io/bionumpy/tutorials/position_weight_matrix.html">bionumpy documentation on PWMs</a>.</p>
<p>Here is an example of a motif defined in the <cite>JASPAR</cite> format. It includes the name of the motif in the first line and
in the subsequent lines the counts of specific nucleotide at the given position are provided.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;MA0080.1     SPI1
A  [    14      4      3     56     56      3 ]
C  [    21      2      0      1      0     18 ]
G  [    19     48     52      0      0     34 ]
T  [     3      3      2      0      1      2 ]
</pre></div>
</div>
<p>To define such motif in the YAML specification, one needs to provide the path to the file where the motif is stored and
the threshold value - when matching the PWM to a sequence later, this is the threshold to consider the sequence as
containing the motif.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_custom_pwm</span><span class="p">:</span><span class="w"> </span><span class="c1"># this will be the identifier of the motif</span>
<span class="w">      </span><span class="nt">file_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_pwm_1.jaspar</span>
<span class="w">      </span><span class="nt">threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
</pre></div>
</div>
</section>
</section>
<section id="position-of-the-motif-in-the-sequence">
<h2>Position of the motif in the sequence<a class="headerlink" href="#position-of-the-motif-in-the-sequence" title="Link to this heading">¶</a></h2>
<p>LIgO supports the use of IMGT positions to specify where the motifs of one signal may occur in the sequence. To specify
the positions for the signal, one can define the position and the corresponding probability. For positions not explicitly
mentioned in the definition, the probability of the motifs occurring will be redistributed from remaining probability.
Specifically, for the example below, the motifs have the probability of 0.9 to occur at positions <cite>109</cite> and <cite>110</cite> and 0.1 total probability
to occur at any other position in the sequence. If some positions are explicitly not allowed, their probability can be
set to 0.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">sequence_position_weights</span><span class="p">:</span><span class="w"> </span><span class="c1"># the motifs have the probability of 0.9 to occur at positions 109 and 110</span>
<span class="w">    </span><span class="s">&#39;109&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.5</span>
<span class="w">    </span><span class="s">&#39;110&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>
</pre></div>
</div>
</section>
<section id="user-defined-functions-for-signal-definition">
<h2>User-defined functions for signal definition<a class="headerlink" href="#user-defined-functions-for-signal-definition" title="Link to this heading">¶</a></h2>
<p>While the previously presented options allow for flexible definitions of motifs and signals, it is possible that the
user might have a different idea of how to define the signal. For that purpose, LIgO supports defining custom functions
that will for the given sequence return True/False based on whether the signal is in the sequence. For more details on
this option, see <a class="reference internal" href="simulation_with_custom_signal_functions.html#simulation-with-custom-signal-functions"><span class="std std-ref">Simulation with custom signal functions</span></a>.</p>
</section>
<section id="complete-example-of-signal-definition-for-receptor-level-simulation">
<h2>Complete example of signal definition for receptor-level simulation<a class="headerlink" href="#complete-example-of-signal-definition-for-receptor-level-simulation" title="Link to this heading">¶</a></h2>
<p>Here is an example of how a set of motifs can be defined and put together under <cite>my_signal</cite>.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">  </span><span class="nt">my_simple_motif</span><span class="p">:</span><span class="w"> </span><span class="c1"># the name of the motif used for reference later</span>
<span class="w">      </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AAA</span><span class="w"> </span><span class="c1"># motif is always AAA</span>
<span class="w">  </span><span class="nt">my_gapped_motif</span><span class="p">:</span><span class="w"> </span><span class="c1"># the name of the more complex motif where / sign denotes a possible gap location</span>
<span class="w">      </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AA/A</span><span class="w"> </span><span class="c1"># this motif can be AAA, AA_A, CAA, CA_A, DAA, DA_A, EAA, EA_A</span>
<span class="w">      </span><span class="nt">min_gap</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span><span class="w"> </span><span class="c1"># how many gaps can there be: min 0 and max 1</span>
<span class="w">      </span><span class="nt">max_gap</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">      </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span><span class="w"> </span><span class="c1"># it can have a max of 1 substitution</span>
<span class="w">          </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span>
<span class="w">          </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.3</span>
<span class="w">      </span><span class="nt">position_weights</span><span class="p">:</span><span class="w"> </span><span class="c1"># note that index 2, the position of the gap, is excluded from position_weights</span>
<span class="w">          </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"> </span><span class="c1"># only first position can be changed</span>
<span class="w">          </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">          </span><span class="nt">3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">      </span><span class="nt">alphabet_weights</span><span class="p">:</span><span class="w"> </span><span class="c1"># the first A can be replaced by C, D or E</span>
<span class="w">          </span><span class="nt">C</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>
<span class="w">          </span><span class="nt">D</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>
<span class="w">          </span><span class="nt">E</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span>
<span class="nt">signals</span><span class="p">:</span>
<span class="w">  </span><span class="nt">my_signal</span><span class="p">:</span><span class="w"> </span><span class="c1"># the name of the signal used for reference later in the simulation specification</span>
<span class="w">      </span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_simple_motif</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_gapped_motif</span>
<span class="w">      </span><span class="nt">sequence_position_weights</span><span class="p">:</span>
<span class="w">          </span><span class="s">&#39;109&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.5</span>
<span class="w">          </span><span class="s">&#39;110&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>
<span class="w">      </span><span class="nt">v_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRBV1</span>
<span class="w">      </span><span class="nt">j_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRBJ1</span>
</pre></div>
</div>
<p>Here <cite>my_signal</cite> has two possible motifs that occur in IMGT positions <cite>109</cite> or <cite>110</cite> with probability 0.9 or in any
other position with probability 0.1, and that have to occur in combination with TRBV1 and TRBJ1 genes.</p>
</section>
<section id="repertoire-level-simulation">
<h2>Repertoire-level simulation<a class="headerlink" href="#repertoire-level-simulation" title="Link to this heading">¶</a></h2>
<p>In addition to the immune signal parameters described above, when defining the immune signal on the repertoire level,
we additionally provide the percentage of the
repertoire containing the given immune signal and the clonal frequency. Clonal frequency is modeled via zeta distribution
function from scipy and it is parameterized by shape parameter of the distribution (called <cite>a</cite> in scipy) and the <cite>loc</cite>
parameter that can be used to shift the distribution. Here is an example:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">signals</span><span class="p">:</span>
<span class="w">  </span><span class="nt">my_signal</span><span class="p">:</span><span class="w"> </span><span class="c1"># same signal as before, with added clonal frequency parameters</span>
<span class="w">      </span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_simple_motif</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_gapped_motif</span>
<span class="w">      </span><span class="nt">sequence_position_weights</span><span class="p">:</span>
<span class="w">          </span><span class="s">&#39;109&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.5</span>
<span class="w">          </span><span class="s">&#39;110&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.5</span>
<span class="w">      </span><span class="nt">v_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRBV1</span>
<span class="w">      </span><span class="nt">j_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRBJ1</span>
<span class="w">      </span><span class="nt">clonal_frequency</span><span class="p">:</span>
<span class="w">          </span><span class="nt">a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="w">          </span><span class="nt">loc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
</pre></div>
</div>
</section>
<section id="simulating-immune-events">
<h2>Simulating immune events<a class="headerlink" href="#simulating-immune-events" title="Link to this heading">¶</a></h2>
<p>We can define one or more signals as described above. If we want to combine multiple signals under a single label to
denote a single immune event (e.g., T1D disease state), LIgO supports this in the following way:</p>
<ol class="arabic simple">
<li><p>First all signals are defined,</p></li>
<li><p>The simulation configuration is provided that defines how the signals will be combined.</p></li>
</ol>
<p>For each group of examples with the same parameters, we define the combination of signals under <cite>signals</cite> key. Then, for
that group, we assign the value of the label of interest. For example, to simulate a group of 30 type 1 diabetes (T1D)-specific
repertoires, where 20 repertoires have the disease and 10 do not, we may specify that positive repertoires contain
<cite>my_signal</cite>, but all of these repertoires contain <cite>my_other_signal</cite>, which could be specific to some other immune event
that is present in the full cohort. An example of such simulation configuration is provided below.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_simulation_config</span><span class="p">:</span>
<span class="w">  </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span><span class="w"> </span><span class="c1"># we do repertoire-level simulation</span>
<span class="w">  </span><span class="nt">paired</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">  </span><span class="nt">sequence_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">amino_acid</span>
<span class="w">  </span><span class="nt">simulation_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">RejectionSampling</span>
<span class="w">  </span><span class="nt">remove_seqs_with_signals</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span><span class="w"> </span><span class="c1"># remove signal-specific AIRs from the background</span>
<span class="w">  </span><span class="nt">sim_items</span><span class="p">:</span>
<span class="w">    </span><span class="nt">t1d_positive_repertoires_group1</span><span class="p">:</span><span class="w"> </span><span class="c1"># group of AIRs with the same parameters</span>
<span class="w">      </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">        </span><span class="nt">chain</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta</span>
<span class="w">        </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">        </span><span class="nt">model_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">null</span>
<span class="w">        </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">      </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span><span class="w"> </span><span class="c1"># we simulate 10 repertoires</span>
<span class="w">      </span><span class="nt">receptors_in_repertoire_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span><span class="w"> </span><span class="c1"># each repertoire has 100 receptors</span>
<span class="w">      </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">        </span><span class="nt">my_signal</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.1</span><span class="w"> </span><span class="c1"># 10% of the receptors in the repertoire contain my_signal</span>
<span class="w">        </span><span class="nt">my_other_signal</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span><span class="w"> </span><span class="c1"># 20% of the receptors contain my_other_signal</span>
<span class="w">      </span><span class="nt">immune_events</span><span class="p">:</span>
<span class="w">        </span><span class="nt">T1D</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">    </span><span class="nt">t1d_positive_repertoires_group2</span><span class="p">:</span><span class="w"> </span><span class="c1"># group of AIRs with the same parameters</span>
<span class="w">      </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">        </span><span class="nt">chain</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta</span>
<span class="w">        </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">        </span><span class="nt">model_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">null</span>
<span class="w">        </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">      </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span><span class="w"> </span><span class="c1"># we simulate 10 repertoires</span>
<span class="w">      </span><span class="nt">receptors_in_repertoire_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span><span class="w"> </span><span class="c1"># each repertoire has 100 receptors</span>
<span class="w">      </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">        </span><span class="nt">my_signal</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.15</span><span class="w"> </span><span class="c1"># 15% of the receptors in the repertoire contain my_signal</span>
<span class="w">        </span><span class="nt">my_other_signal</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span><span class="w"> </span><span class="c1"># 20% of the receptors contain my_other_signal</span>
<span class="w">      </span><span class="nt">immune_events</span><span class="p">:</span>
<span class="w">        </span><span class="nt">T1D</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">    </span><span class="nt">t1d_negative_receptors</span><span class="p">:</span>
<span class="w">      </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">        </span><span class="nt">chain</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta</span>
<span class="w">        </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">        </span><span class="nt">model_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">null</span>
<span class="w">        </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">      </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span><span class="w"> </span><span class="c1"># we simulate 10 repertoires</span>
<span class="w">      </span><span class="nt">receptors_in_repertoire_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span><span class="w"> </span><span class="c1"># each repertoire has 100 receptors</span>
<span class="w">      </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">        </span><span class="nt">my_other_signal</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.03</span><span class="w"> </span><span class="c1"># 3% of the receptors contain my_other_signal but none contain my_signal</span>
<span class="w">      </span><span class="nt">immune_events</span><span class="p">:</span>
<span class="w">        </span><span class="nt">T1D</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
</pre></div>
</div>
<section id="immune-events-on-the-receptor-level">
<h3>Immune events on the receptor level<a class="headerlink" href="#immune-events-on-the-receptor-level" title="Link to this heading">¶</a></h3>
<p>It is possible to define immune events in the same way on the receptor level. They could denote the same as on the
repertoire level: the disease state to which the signals listed are specific to, or some other label of interest, e.g.,
experiment or patient from which the receptor came from.</p>
</section>
</section>
<section id="next-steps">
<h2>Next steps<a class="headerlink" href="#next-steps" title="Link to this heading">¶</a></h2>
<p>For a full minimal working example, see the <a class="reference internal" href="../quickstart.html#quickstart"><span class="std std-ref">Quickstart</span></a>. For detailed description of all the parameters and possible
values, see <a class="reference internal" href="../specification.html#yaml-specification"><span class="std std-ref">YAML specification</span></a>. For help on choosing the content of the signals and motifs, see <a class="reference internal" href="how_to_check_feasibility_of_sim_params.html#how-to-check-feasibility-of-the-simulation-parameters"><span class="std std-ref">How to check feasibility of the simulation parameters</span></a>.
For the discussion on defining immune events and signals in this way, see also <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2023.10.20.562936v2">the LIgO manuscript</a>.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="how_to_choose_simulation_strategy.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">How to choose between rejection sampling and signal implantation</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../tutorials.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorials</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">How to define immune signals and immune events</a><ul>
<li><a class="reference internal" href="#motifs">Motifs</a><ul>
<li><a class="reference internal" href="#seed-motif">Seed motif</a></li>
<li><a class="reference internal" href="#positional-weight-matrix">Positional weight matrix</a></li>
</ul>
</li>
<li><a class="reference internal" href="#position-of-the-motif-in-the-sequence">Position of the motif in the sequence</a></li>
<li><a class="reference internal" href="#user-defined-functions-for-signal-definition">User-defined functions for signal definition</a></li>
<li><a class="reference internal" href="#complete-example-of-signal-definition-for-receptor-level-simulation">Complete example of signal definition for receptor-level simulation</a></li>
<li><a class="reference internal" href="#repertoire-level-simulation">Repertoire-level simulation</a></li>
<li><a class="reference internal" href="#simulating-immune-events">Simulating immune events</a><ul>
<li><a class="reference internal" href="#immune-events-on-the-receptor-level">Immune events on the receptor level</a></li>
</ul>
</li>
<li><a class="reference internal" href="#next-steps">Next steps</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=9306f2a9"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>