<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Simulating epitope-specific TCRs and validating their properties" href="epitope-specific_TCR_AIRRs.html" /><link rel="prev" title="How to simulate co-occuring immune signals" href="how_to_simulate_co-occuring_signals.html" />
        <link rel="canonical" href="https://uio-bmi.github.io/ligo/tutorials/TCR_motifs.html" />

    <link rel="shortcut icon" href="../_static/favicon.ico"/><!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities - LIgO 1.0.11 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/immuneml.css?v=3985a9e7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">LIgO 1.0.11 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">LIgO 1.0.11 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 has-children"><a class="reference internal" href="../quickstart.html">Quickstart</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Quickstart</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/receptor_level_simulation.html">How to use LIgO for receptor-level simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/repertoire_level_simulation.html">How to use LIgO for repertoire-level simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/overview_of_ligo_parameters.html">Overview of LIgO simulation parameters</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../installation.html">Installing LIgO</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Installing LIgO</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="how_to_define_immune_signals_and_events.html">How to define immune signals and immune events</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_choose_simulation_strategy.html">How to choose between rejection sampling and signal implantation</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_check_feasibility_of_sim_params.html">How to check feasibility of the simulation parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_handle_the_maximum_iterations_were_reached_error.html">How to handle the maximum iterations were reached error</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulation_with_custom_signal_functions.html">Simulation with custom signal functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_simulate_paired_chain_data.html">Paired chain simulations in LIgO</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_simulate_co-occuring_signals.html">How to simulate co-occuring immune signals</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</a></li>
<li class="toctree-l2"><a class="reference internal" href="epitope-specific_TCR_AIRRs.html">Simulating epitope-specific TCRs and validating their properties</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../usecases.html">Use case examples</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Use case examples</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../usecases/usecase1.html">Manuscript use case 1: Out-of-distribution receptor-level simulation using LIgO</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/usecase2.html">Manuscript use case 2: Limitations of conventional encoding schemes for repertoire-level binary classification when immune signals co-occur within the same AIR</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../specification.html">YAML specification</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of YAML specification</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/tutorials/TCR_motifs.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="constructing-ligo-motifs-inspired-by-a-database-of-tcr-sequences-with-known-antigen-specificities">
<h1>Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities<a class="headerlink" href="#constructing-ligo-motifs-inspired-by-a-database-of-tcr-sequences-with-known-antigen-specificities" title="Link to this heading">¶</a></h1>
<p>LIgO enables the generation of motifs using either (i) a combination of a short amino acid sequence (i.e., a seed) and a list of Hamming distances, representing the allowed deviations from the seed or (ii) a position weight matrix (PWM).</p>
<p>In this tutorial, we demonstrate how one can use a database of TCRs with known antigen specificities to construct more biologically-relevant LIgO motifs. These database-inspired motifs can be used to simulate epitope-specific TCR data, i.e., a collection of TCRs derived from different individuals that are all recognizing the same epitope.</p>
<p>We suggest two approaches for constructing database-inspired LIgO motifs:</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="#basic-approach-defining-motif-based-on-a-seed-and-a-list-of-hamming-distances"><span class="std std-ref">Basic approach: defining motif based on a seed and a list of Hamming distances</span></a></p></li>
<li><p><a class="reference internal" href="#enhanced-approach-defining-a-motif-based-on-pwm"><span class="std std-ref">Enhanced approach: defining a motif based on PWM</span></a></p></li>
</ol>
<p>In both approaches we will use the  <a class="reference external" href="https://vdjdb.cdr3.net/">VDJdb database</a>  (Shugay et al., 2018; Goncharov et al., 2022) to obtain epitope-specific TCRs. However, there are more databases available, such as</p>
<ul class="simple">
<li><p><a class="reference external" href="https://friedmanlab.weizmann.ac.il/McPAS-TCR/">McPAS</a> (Tickotsky et al., 2017)</p></li>
<li><p><a class="reference external" href="https://www.iedb.org/">IEDB</a>  (Vita et al., 2019)</p></li>
<li><p><a class="reference external" href="https://atlas.wenglab.org/web/">ATLAS</a>  (Borrman et al., 2017)</p></li>
<li><p><a class="reference external" href="http://bioinfo.life.hust.edu.cn/TCRdb/#/">TCRdb</a>  (Chen et al., 2021)</p></li>
</ul>
<p>Since similar approach can be applied to the BCR simulation, we also provide BCR/antibody databases:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://opig.stats.ox.ac.uk/webapps/covabdab/">CoV-AbDab</a> (Raybould et al., 2021)</p></li>
<li><p><a class="reference external" href="https://opig.stats.ox.ac.uk/webapps/plabdab/">PLAbDab</a> (Abanades et al., 2024)</p></li>
<li><p><a class="reference external" href="https://opig.stats.ox.ac.uk/webapps/oas">OAS</a> (Olsen, 2022)</p></li>
</ul>
<section id="basic-approach-defining-motif-based-on-a-seed-and-a-list-of-hamming-distances">
<h2>Basic approach: defining motif based on a seed and a list of Hamming distances<a class="headerlink" href="#basic-approach-defining-motif-based-on-a-seed-and-a-list-of-hamming-distances" title="Link to this heading">¶</a></h2>
<section id="step-1-constructing-seeds-from-the-vdjdb-receptors">
<h3>Step 1: Constructing seeds from the VDJdb receptors<a class="headerlink" href="#step-1-constructing-seeds-from-the-vdjdb-receptors" title="Link to this heading">¶</a></h3>
<p>For demonstration purposes, we define three seeds from three TCR beta sequences in the VDJdb (Shugay et al. 2018), all recognizing the DENV3/4 epitope GTSGSPIINR (Table 1). These TCRs need to be transformed to become LIgO seeds. Typically, the epitope-binding motif is located in the middle of CDR3 beta sequences, while the beginning and end are responsible for HLA-binding. Therefore, the center of the CDR3 beta sequence can be used as seed. In this tutorial, we will consider two ways of constructing seeds: <em>long seeds</em> and <em>short seeds</em>.</p>
<ul class="simple">
<li><p>To construct <em>long seeds</em>, we removed the first and the last 3 amino acids from the CDR3 sequence;</p></li>
<li><p>To construct <em>short seeds</em>, we randomly selected a consecutive substring of 3 amino acid long out of the long seed sequence in such way that the resulting seeds are not too similar.</p></li>
</ul>
<p>Table 1 shows the long and short seeds that were selected for the three initial TCRs. We will use these two seed sets to perform two LIgO simulations, keeping all other simulation parameters identical except for the seeds.</p>
<div class="table-wrapper docutils container" id="id2">
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">Table 1: Examples of the long and short seeds</span><a class="headerlink" href="#id2" title="Link to this table">¶</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>TCR beta sequence</p></th>
<th class="head"><p>TRBV gene</p></th>
<th class="head"><p>TRBJ gene</p></th>
<th class="head"><p>Epitope</p></th>
<th class="head"><p>Long Seed</p></th>
<th class="head"><p>Short Seed</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CSVELSGINQPQHF</p></td>
<td><p>TRBV29-1</p></td>
<td><p>TRBJ1-5</p></td>
<td><p>GTSGSPIINR</p></td>
<td><p>ELSGINQP</p></td>
<td><p>SGI</p></td>
</tr>
<tr class="row-odd"><td><p>CASSPAGGTYEQYF</p></td>
<td><p>TRBV11-2</p></td>
<td><p>TRBJ2-7</p></td>
<td><p>GTSGSPIINR</p></td>
<td><p>SPAGGTYE</p></td>
<td><p>PAG</p></td>
</tr>
<tr class="row-even"><td><p>CASSGGDVREEQYF</p></td>
<td><p>TRBV9</p></td>
<td><p>TRBJ2-7</p></td>
<td><p>GTSGSPIINR</p></td>
<td><p>SGGDVREE</p></td>
<td><p>DVR</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="step-2-defining-ligo-motifs-based-on-the-seeds">
<h3>Step 2: Defining LIgO motifs based on the seeds<a class="headerlink" href="#step-2-defining-ligo-motifs-based-on-the-seeds" title="Link to this heading">¶</a></h3>
<p>After describing the seeds, it is required to define possible deviations that are allowed from these seeds using hamming distance. To select the maximum Hamming distance, consider the length of your seed and the aimed diversity of the final TCR repertoire.</p>
<p>Shorter seeds require lower Hamming distances. However, if the user only wishes to simulated TCRs looking very similar to the seed, the user may also lower the hamming distance.</p>
<p>In this tutorial, a maximum Hamming distance of two was selected to restrict the diversity of the simulated epitope-specific TCR receptors. Below we show an example of how to define motifs using Hamming distance and long seeds. These signals can be further used for LIgO simulation based on rejection sampling or signal implantation.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">   </span><span class="nt">motif1</span><span class="p">:</span>
<span class="w">     </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span>
<span class="w">       </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.1</span><span class="w"> </span><span class="c1"># 10% of TCRs will contain ELSGINQP as the exact match</span>
<span class="w">       </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span><span class="w"> </span><span class="c1"># 20% of TCRs will contain ELSGINQP with 1 mismatch</span>
<span class="w">       </span><span class="nt">2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span><span class="w"> </span><span class="c1"># 70% of TCRs will contain ELSGINQP with 2 mismatch</span>
<span class="w">     </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ELSGINQP</span>
<span class="w">   </span><span class="nt">motif2</span><span class="p">:</span>
<span class="w">     </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span>
<span class="w">       </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.1</span><span class="w"> </span><span class="c1"># 10% of TCRs will contain SPAGGTYE as the exact match</span>
<span class="w">       </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span><span class="w"> </span><span class="c1"># 20% of TCRs will contain SPAGGTYE with 1 mismatch</span>
<span class="w">       </span><span class="nt">2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span><span class="w"> </span><span class="c1"># 70% of TCRs will contain SPAGGTYE with 2 mismatch</span>
<span class="w">     </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">SPAGGTYE</span>
<span class="w">   </span><span class="nt">motif3</span><span class="p">:</span>
<span class="w">     </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span>
<span class="w">       </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.1</span><span class="w"> </span><span class="c1"># 10% of TCRs will contain SGGDVREE as the exact match</span>
<span class="w">       </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span><span class="w"> </span><span class="c1"># 20% of TCRs will contain SGGDVREE with 1 mismatch</span>
<span class="w">       </span><span class="nt">2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span><span class="w"> </span><span class="c1"># 70% of TCRs will contain SGGDVREE with 2 mismatch</span>
<span class="w">     </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">SGGDVREE</span>
</pre></div>
</div>
</section>
<section id="general-tips-for-defining-a-motif-using-a-seed-and-hamming-distance">
<h3>General tips for defining a motif using a seed and Hamming distance<a class="headerlink" href="#general-tips-for-defining-a-motif-using-a-seed-and-hamming-distance" title="Link to this heading">¶</a></h3>
<ol class="arabic">
<li><p>Start with the full seed you want to find back in your simulated TCRs, e.g., ELSGINQP</p></li>
<li><p>If you want to use rejection sampling, estimate the maximal hamming distance to finish your simulation in a reasonable time. You can start with a very restrictive hamming distance (e.g. max 1) and adjust it as needed. You can use the feasibility report to estimate the effectiveness of the simulation with a given set of parameters, see <a class="reference internal" href="how_to_check_feasibility_of_sim_params.html#how-to-check-feasibility-of-the-simulation-parameters"><span class="std std-ref">How to check feasibility of the simulation parameters</span></a>.</p>
<p>For example, in this tutorial we used the following rule of thumb:</p>
<ul class="simple">
<li><p>Seed length of 6-8 =&gt; max Hamming distance = 2</p></li>
<li><p>Seed length of 9-10 =&gt; max Hamming distance = 3</p></li>
<li><p>Seed length &gt;10 =&gt; test the simulation with a maximal Hamming distance of 3. If not enough TCR are simulated, increase the max Hamming distance up to 4.</p></li>
</ul>
<p>If you want to use implanting, you do not need to estimate the feasibility because the simulation will be fast with any Hamming distance.</p>
</li>
<li><p>Start the simulation with the selected seed and Hamming distances. Check for the presence of the predefined motif in the simulated TCRs by clustering or allocating the seed within the TCR sequences.</p></li>
</ol>
</section>
</section>
<section id="enhanced-approach-defining-a-motif-based-on-pwm">
<h2>Enhanced approach: defining a motif based on PWM<a class="headerlink" href="#enhanced-approach-defining-a-motif-based-on-pwm" title="Link to this heading">¶</a></h2>
<section id="option-1-vdjdb-cdr3-motif-database">
<h3>Option 1: VDJdb CDR3 motif database<a class="headerlink" href="#option-1-vdjdb-cdr3-motif-database" title="Link to this heading">¶</a></h3>
<p>VDJdb provides a database of CDR3 motifs, which you can access at <a class="reference external" href="https://vdjdb.cdr3.net/motif">https://vdjdb.cdr3.net/motif</a> to find suitable motifs. You can search for motifs based on an epitope or a CDR3 sequence or subsequence.</p>
<p>The screenshot below showcases a SARS-CoV-2-specific motif from VDJdb. These SARS-CoV-2-specific TCRs must have 15 amino acids long CDR3 and use the germline genes TRBV27*01 and TRBJ2-1*1.</p>
<a class="reference internal image-reference" href="../_images/VDJdb_CDR3_database.png"><img alt="../_images/VDJdb_CDR3_database.png" src="../_images/VDJdb_CDR3_database.png" style="width: 1500px;" />
</a>
<p>After clicking the Export button, you will receive a TSV file containing all the members representing the given motif. To convert the set of CDR3s into a PWM compatible with LIgO, you can use the following code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;ClusterMembers_H.A.IVTDFSVIK.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;cdr3aa&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
<span class="n">alignment</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">construct_pwm</span><span class="p">(</span><span class="n">alignment</span><span class="p">):</span>
    <span class="n">amino_acids</span> <span class="o">=</span> <span class="s1">&#39;ACDEFGHIKLMNPQRSTVWY&#39;</span>
    <span class="n">aa_to_index</span> <span class="o">=</span> <span class="p">{</span><span class="n">aa</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">amino_acids</span><span class="p">)}</span>
    <span class="n">alignment_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">aa_to_index</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">]</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment</span><span class="p">])</span>
    <span class="n">pwm_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
        <span class="n">np</span><span class="o">.</span><span class="n">bincount</span><span class="p">(</span><span class="n">alignment_matrix</span><span class="p">[:,</span> <span class="n">pos</span><span class="p">][</span><span class="n">alignment_matrix</span><span class="p">[:,</span> <span class="n">pos</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">],</span> <span class="n">minlength</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">amino_acids</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">get_alignment_length</span><span class="p">())</span>
        <span class="p">])</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
    <span class="n">pwm_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pwm</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">amino_acids</span><span class="p">))</span>
    <span class="k">return</span><span class="p">(</span><span class="n">pwm_df</span><span class="p">)</span>

<span class="n">pwm</span> <span class="o">=</span> <span class="n">construct_pwm</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="n">pwm</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;pwm.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, the motif file pwm.csv can be used to define LIgO signal. Additional CDR3 length restricrion (15AA) can be added using the sequence_len_limits parameter in the simulation config item</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">  </span><span class="nt">sars-cov-2_motif</span><span class="p">:</span>
<span class="w">    </span><span class="nt">file_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pwm.csv</span>
<span class="w">       </span><span class="l l-Scalar l-Scalar-Plain">threshold</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span><span class="w"> </span><span class="c1"># the threshold to consider the sequence as containing the motif, can be changed</span>
<span class="nt">signals</span><span class="p">:</span>
<span class="w">   </span><span class="nt">signal1</span><span class="p">:</span>
<span class="w">     </span><span class="nt">v_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRBV27*01</span>
<span class="w">     </span><span class="nt">j_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRBJ2-1*01</span>
<span class="w">     </span><span class="nt">motifs</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">sars-cov-2_motif</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
</section>
<section id="option-2-database-clustering-by-the-user">
<h3>Option 2: Database clustering by the user<a class="headerlink" href="#option-2-database-clustering-by-the-user" title="Link to this heading">¶</a></h3>
<p>Alternatively, one can cluster VDJdb (or any other database) to obtain PWMs for LIgO simulation. In this tutorial we demonstrate how to cluster VDJdb database using the clusTCR tool (Valkiers et al., 2021). For more information on clusTCR, see <a class="reference external" href="https://svalkiers.github.io/clusTCR/">clusTCR documentation</a>.</p>
<p>First, one should select a set of TCRs from VDJdb sharing the same epitope specificity (see <a class="reference external" href="https://vdjdb.cdr3.net/search">vdjdb.cdr3.net/search</a>). To obtain accurate PWMs, consider epitope sequences containing a sufficient number of epitope-specific TCRs. For demonstration purposes, we used InfluenzaA epitope GILGFVFTL which is associated with 9490 TCRB receptors, see the figure below.</p>
<a class="reference internal image-reference" href="../_images/VDJdb_download.png"><img alt="../_images/VDJdb_download.png" src="../_images/VDJdb_download.png" style="width: 1500px;" />
</a>
<p>Next, run the following script on the downloaded VDJdb data to cluster the TCR receptors and generate clusTCR motifs:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">clustcr</span> <span class="kn">import</span> <span class="n">Clustering</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Read downloaded vdjdb data</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;vdjdb.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># Remove duplicated CDR3 beta sequences</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s1">&#39;CDR3&#39;</span><span class="p">)</span>

<span class="c1"># Cluster training data for selected epitope using clusTCR</span>
<span class="n">clustering</span> <span class="o">=</span> <span class="n">Clustering</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;MCL&#39;</span><span class="p">)</span>
<span class="n">clustered_data</span> <span class="o">=</span> <span class="n">clustering</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;CDR3&#39;</span><span class="p">])</span>

<span class="c1"># Save clusTCR motifs sorted by cluster size</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">clustered_data</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">motifs</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;size&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">motifs</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;clustcr_motifs.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>The clustcr_motifs.csv file will contain motifs saved in clusTCR format, see the example below. You can read more about clusTCR motif format in the <a class="reference external" href="https://svalkiers.github.io/clusTCR/docs/clustering/how-to-use.html#summary">clusTCR documentation</a>. Briefly, the motif uses upper-case for highly conserved amino acids (frequency &gt; 0.7) and lower-case for moderately conserved ones. If two amino acids are equally frequent, they are in brackets ([ ]), and less significant positions use a dot (.) as a wildcard.</p>
<p>One can use clusTCR motifs to define LIgO PWM motifs. Alternatively, PWM can be constructed based on the TCRs belonging to the same cluster, using the <code class="code docutils literal notranslate"><span class="pre">construct_pwm</span></code> function defined earlier in the tutorial. To obtain the TCRs belonging to each cluster use <code class="code docutils literal notranslate"><span class="pre">clustered_data.clusters_df</span></code>.</p>
<div class="table-wrapper docutils container" id="id3">
<table class="docutils align-default" id="id3">
<caption><span class="caption-text">Table 2: ClusTCR motifs of top-3 largest clusters constructed using GILGFVFTL-specific TCRs</span><a class="headerlink" href="#id3" title="Link to this table">¶</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>size</p></th>
<th class="head"><p>motif</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>361</p></td>
<td><p>CASS.Rs..EQyF</p></td>
</tr>
<tr class="row-odd"><td><p>199</p></td>
<td><p>ASS..s.DTQYF</p></td>
</tr>
<tr class="row-even"><td><p>124</p></td>
<td><p>CASS..SnQPQHF</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="epitope-specific_TCR_AIRRs.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Simulating epitope-specific TCRs and validating their properties</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="how_to_simulate_co-occuring_signals.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">How to simulate co-occuring immune signals</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</a><ul>
<li><a class="reference internal" href="#basic-approach-defining-motif-based-on-a-seed-and-a-list-of-hamming-distances">Basic approach: defining motif based on a seed and a list of Hamming distances</a><ul>
<li><a class="reference internal" href="#step-1-constructing-seeds-from-the-vdjdb-receptors">Step 1: Constructing seeds from the VDJdb receptors</a></li>
<li><a class="reference internal" href="#step-2-defining-ligo-motifs-based-on-the-seeds">Step 2: Defining LIgO motifs based on the seeds</a></li>
<li><a class="reference internal" href="#general-tips-for-defining-a-motif-using-a-seed-and-hamming-distance">General tips for defining a motif using a seed and Hamming distance</a></li>
</ul>
</li>
<li><a class="reference internal" href="#enhanced-approach-defining-a-motif-based-on-pwm">Enhanced approach: defining a motif based on PWM</a><ul>
<li><a class="reference internal" href="#option-1-vdjdb-cdr3-motif-database">Option 1: VDJdb CDR3 motif database</a></li>
<li><a class="reference internal" href="#option-2-database-clustering-by-the-user">Option 2: Database clustering by the user</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=9306f2a9"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>