<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="prev" title="Manuscript use case 2: Limitations of conventional encoding schemes for repertoire-level binary classification when immune signals co-occur within the same AIR" href="usecases/usecase2.html" />
        <link rel="canonical" href="https://uio-bmi.github.io/ligo/specification.html" />

    <link rel="shortcut icon" href="_static/favicon.ico"/><!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>YAML specification - LIgO 1.0.11 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="_static/css/immuneml.css?v=3985a9e7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">LIgO 1.0.11 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">LIgO 1.0.11 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 has-children"><a class="reference internal" href="quickstart.html">Quickstart</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Quickstart</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="quickstart/receptor_level_simulation.html">How to use LIgO for receptor-level simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart/repertoire_level_simulation.html">How to use LIgO for repertoire-level simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="quickstart/overview_of_ligo_parameters.html">Overview of LIgO simulation parameters</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="installation.html">Installing LIgO</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Installing LIgO</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="tutorials.html">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorials/how_to_define_immune_signals_and_events.html">How to define immune signals and immune events</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/how_to_choose_simulation_strategy.html">How to choose between rejection sampling and signal implantation</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/how_to_check_feasibility_of_sim_params.html">How to check feasibility of the simulation parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/how_to_handle_the_maximum_iterations_were_reached_error.html">How to handle the maximum iterations were reached error</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/simulation_with_custom_signal_functions.html">Simulation with custom signal functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/how_to_simulate_paired_chain_data.html">Paired chain simulations in LIgO</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/how_to_simulate_co-occuring_signals.html">How to simulate co-occuring immune signals</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/TCR_motifs.html">Constructing LIgO motifs inspired by a database of TCR sequences with known antigen specificities</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/epitope-specific_TCR_AIRRs.html">Simulating epitope-specific TCRs and validating their properties</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="usecases.html">Use case examples</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Use case examples</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="usecases/usecase1.html">Manuscript use case 1: Out-of-distribution receptor-level simulation using LIgO</a></li>
<li class="toctree-l2"><a class="reference internal" href="usecases/usecase2.html">Manuscript use case 2: Limitations of conventional encoding schemes for repertoire-level binary classification when immune signals co-occur within the same AIR</a></li>
</ul>
</li>
<li class="toctree-l1 current has-children current-page"><a class="current reference internal" href="#">YAML specification</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of YAML specification</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="_sources/specification.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="yaml-specification">
<h1>YAML specification<a class="headerlink" href="#yaml-specification" title="Link to this heading">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<p>The YAML specification defines the simulation or analysis to be executed. The YAML contains the following sections:</p>
<ul class="simple">
<li><p>definitions: includes motifs, signals, and simulation details - these define conceptually what the user would like to simulate,</p></li>
<li><p>instructions: cover the specific simulation or analysis to be run - usually including only technical information on how to perform the task and reference the components defined under definitions for the conceptual information;</p></li>
<li><p>output: only HTML format is supported, and doesn’t include any parameters.</p></li>
</ul>
<p>The purpose of this page is to list all the YAML specification options.</p>
<p>The overall structure of the YAML specification is the following:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">definitions</span><span class="p">:</span><span class="w"> </span><span class="c1"># mandatory keyword</span>
<span class="w">  </span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_motif_1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">...</span>
<span class="w">  </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_sig_1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">...</span>
<span class="w">  </span><span class="nt">simulations</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_first_simulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">...</span>
<span class="nt">instructions</span><span class="p">:</span><span class="w"> </span><span class="c1"># mandatory keyword - at least one instruction has to be specified</span>
<span class="w">  </span><span class="nt">my_instruction_1</span><span class="p">:</span><span class="w"> </span><span class="c1"># user-defined name of the instruction</span>
<span class="w">    </span><span class="l l-Scalar l-Scalar-Plain">...</span><span class="w"> </span><span class="c1"># see below for the specification of different instructions</span>
<span class="nt">output</span><span class="p">:</span><span class="w"> </span><span class="c1"># how to present the result after running (the only valid option now is HTML)</span>
<span class="w">  </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">HTML</span>
</pre></div>
</div>
<section id="definitions">
<h2>Definitions<a class="headerlink" href="#definitions" title="Link to this heading">¶</a></h2>
<section id="simulation">
<h3>Simulation<a class="headerlink" href="#simulation" title="Link to this heading">¶</a></h3>
<section id="seedmotif">
<h4>SeedMotif<a class="headerlink" href="#seedmotif" title="Link to this heading">¶</a></h4>
<p>Describes motifs by seed, possible gaps, allowed hamming distances, positions that can be changed and what they can be changed to.</p>
<p>Arguments:</p>
<ul class="simple">
<li><p>seed (str): An amino acid or nucleotide sequence that represents the basic motif seed. The forward slash (/) symbol corresponds to a gap position (only one gap per seed is allowed). All implanted motifs correspond to the seed, or a modified version thereof, as specified in its instantiation strategy. If this argument is set, seed_chain1 and seed_chain2 arguments are not used.</p></li>
<li><p>min_gap (int): The minimum gap length, in case the original seed contains a gap.</p></li>
<li><p>max_gap (int): The maximum gap length, in case the original seed contains a gap.</p></li>
<li><p>hamming_distance_probabilities (dict): A distribution describing the probability of the given seed being modified with a certain number of random mutations, which we refer to as modifications. The keys represent the number of modifications (hamming distance) between the original seed and the implanted motif, and the values represent the probabilities for the respective number of modifications. For example, {0: 0.7, 1: 0.3} means that if signal implanting strategy is used, 30% of the time one position of the seed will be modified, and the remaining 70% of the time the motif will remain unmodified with respect to the seed. The values of hamming_distance_probabilities must sum to 1. If signal-specific receptors are generated with rejection sampling strategy, then any sequence containing the seed with the allowed number of mismatches (non-zero probabilities) will be selected.</p></li>
<li><p>position_weights (dict): A distribution across seed positions describing probabilities of modifying each seed position. The keys represent the position in the seed, where counting starts at 0. If the index of a gap is specified in position_weights, it will be removed. The values represent the relative probabilities for modifying each position when it gets selected for modification. For example {0: 0.6, 1: 0, 2: 0.4} means that if signal implanting strategy is used, when a sequence is selected for a modification (as specified in hamming_distance_probabilities), then 60% of the time the amino acid at index 0 is modified, and the remaining 40% of the time the amino acid at index 2. If the values of position_weights do not sum to 1, the remainder will be redistributed over all positions, including those not specified. If signal-specific receptors are generated with rejection sampling strategy, then modifications are only allowed at the seed positions with non-zero probabilities.</p></li>
<li><p>alphabet_weights (dict): A distribution across the nucleotide or amino acid alphabet describing the probability of a given nucleotide or amino acid to be used for modification. The keys of the dictionary represent the amino acids and the values are the relative probabilities for choosing this amino acid. If the values of alphabet_weights do not sum to 1, the remainder will be redistributed over all possible amino acids, including those not specified. If signal-specific receptors are generated with rejection sampling strategy, then any sequence containing the seed with modifications having non-zero probabilities will be selected.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">    </span><span class="c1"># examples for single chain receptor data</span>
<span class="w">    </span><span class="nt">my_simple_motif</span><span class="p">:</span><span class="w"> </span><span class="c1"># this will be the identifier of the motif</span>
<span class="w">        </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AAA</span><span class="w"> </span><span class="c1"># motif is always AAA</span>
<span class="w">    </span><span class="nt">my_gapped_motif</span><span class="p">:</span>
<span class="w">        </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AA/A</span><span class="w"> </span><span class="c1"># this motif can be AAA, AA_A, CAA, CA_A, DAA, DA_A, EAA, EA_A</span>
<span class="w">        </span><span class="nt">min_gap</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">        </span><span class="nt">max_gap</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">        </span><span class="nt">hamming_distance_probabilities</span><span class="p">:</span><span class="w"> </span><span class="c1"># it can have a max of 1 substitution</span>
<span class="w">            </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span>
<span class="w">            </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.3</span>
<span class="w">        </span><span class="nt">position_weights</span><span class="p">:</span><span class="w"> </span><span class="c1"># note that index 2, the position of the gap, is excluded from position_weights</span>
<span class="w">            </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"> </span><span class="c1"># only first position can be changed</span>
<span class="w">            </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">            </span><span class="nt">3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">        </span><span class="nt">alphabet_weights</span><span class="p">:</span><span class="w"> </span><span class="c1"># the first A can be replaced by C, D or E</span>
<span class="w">            </span><span class="nt">C</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>
<span class="w">            </span><span class="nt">D</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>
<span class="w">            </span><span class="nt">E</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span>
</pre></div>
</div>
</section>
<section id="pwm">
<h4>PWM<a class="headerlink" href="#pwm" title="Link to this heading">¶</a></h4>
<p>Motifs defined by a positional weight matrix and using bionumpy’s PWM internally.
For more details on bionumpy’s implementation of PWM, as well as for supported formats,
see the documentation at <a class="reference external" href="https://bionumpy.github.io/bionumpy/tutorials/position_weight_matrix.html">https://bionumpy.github.io/bionumpy/tutorials/position_weight_matrix.html</a>.</p>
<p>Arguments:</p>
<ul class="simple">
<li><p>file_path: path to the file where the PWM is stored</p></li>
<li><p>threshold (float): when matching PWM to a sequence, this is the threshold to consider the sequence as containing the motif</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_custom_pwm</span><span class="p">:</span><span class="w"> </span><span class="c1"># this will be the identifier of the motif</span>
<span class="w">        </span><span class="nt">file_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_pwm_1.jaspar</span>
<span class="w">        </span><span class="nt">threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
</pre></div>
</div>
</section>
<section id="signal">
<h4>Signal<a class="headerlink" href="#signal" title="Link to this heading">¶</a></h4>
<p>This class represents the simulated signal.
A signal is represented by a list of motifs, and optionally, position weights showing where one of the motifs of the signal can
occur in a sequence.</p>
<p>A signal is associated with a metadata label, which is assigned to a receptor or repertoire.
For example antigen-specific/disease-associated (receptor) or diseased (repertoire).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>IMGT positions</p>
<p>To use sequence position weights, IMGT positions should be explicitly specified as strings, under quotation marks, to allow for all positions to be properly distinguished.</p>
</div>
<p>Arguments:</p>
<ul class="simple">
<li><p>motifs (list): A list of the motifs associated with this signal, either defined by seed or by position weight matrix. Alternatively, it can be a list of a list of motifs, in which case the motifs in the same sublist (max 2 motifs) have to co-occur in the same sequence</p></li>
<li><p>sequence_position_weights (dict): a dictionary specifying for each IMGT position in the sequence how likely it is for the signal to be there. If the position is not present in the sequence, the probability of the signal occurring at that position will be redistributed to other positions with probabilities that are not explicitly set to 0 by the user.</p></li>
<li><p>v_call (str): V gene with allele if available that has to co-occur with one of the motifs for the signal to exist; can be used in combination with rejection sampling, or full sequence implanting, otherwise ignored; to match in a sequence for rejection sampling, it is checked if this value is contained in the same field of generated sequence;</p></li>
<li><p>j_call (str): J gene with allele if available that has to co-occur with one of the motifs for the signal to exist; can be used in combination with rejection sampling, or full sequence implanting, otherwise ignored; to match in a sequence for rejection sampling, it is checked if this value is contained in the same field of generated sequence;</p></li>
<li><p>source_file (str): path to the file where the custom signal function is; cannot be combined with the arguments listed above (motifs, v_call, j_call, sequence_position_weights)</p></li>
<li><p>is_present_func (str): name of the function from the source_file file that will be used to specify the signal; the function’s signature must be:</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">is_present</span><span class="p">(</span><span class="n">sequence_aa</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">v_call</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">j_call</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
    <span class="c1"># custom implementation where all or some of these arguments can be used</span>
</pre></div>
</div>
<ul class="simple">
<li><p>clonal_frequency (dict): clonal frequency in Ligo is simulated through <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.zipf.html">scipy’s zeta distribution function for generating random numbers</a>, with parameters provided under clonal_frequency parameter. If clonal frequency should not be used, this parameter can be None</p></li>
</ul>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">clonal_frequency</span><span class="p">:</span>
<span class="w">  </span><span class="nt">a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span><span class="w"> </span><span class="c1"># shape parameter of the distribution</span>
<span class="w">  </span><span class="nt">loc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span><span class="w"> </span><span class="c1"># 0 by default but can be used to shift the distribution</span>
</pre></div>
</div>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">signals</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_signal</span><span class="p">:</span>
<span class="w">        </span><span class="nt">motifs</span><span class="p">:</span>
<span class="w">            </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_simple_motif</span>
<span class="w">            </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_gapped_motif</span>
<span class="w">        </span><span class="nt">sequence_position_weights</span><span class="p">:</span>
<span class="w">            </span><span class="s">&#39;109&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.5</span>
<span class="w">            </span><span class="s">&#39;110&#39;</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.5</span>
<span class="w">        </span><span class="nt">v_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRBV1</span>
<span class="w">        </span><span class="nt">j_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRBJ1</span>
<span class="w">        </span><span class="nt">clonal_frequency</span><span class="p">:</span>
<span class="w">            </span><span class="nt">a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="w">            </span><span class="nt">loc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">    </span><span class="nt">signal_with_custom_func</span><span class="p">:</span>
<span class="w">        </span><span class="nt">source_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">signal_func.py</span>
<span class="w">        </span><span class="nt">is_present_func</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">is_signal_present</span>
<span class="w">        </span><span class="nt">clonal_frequency</span><span class="p">:</span>
<span class="w">            </span><span class="nt">a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="w">            </span><span class="nt">loc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
</pre></div>
</div>
</section>
<section id="simulation-config">
<h4>Simulation config<a class="headerlink" href="#simulation-config" title="Link to this heading">¶</a></h4>
<p>Defines all parameters of the simulation.</p>
<p>Arguments:</p>
<ul class="simple">
<li><p>sim_items (dict): a list of SimConfigItems defining individual units of simulation</p></li>
<li><p>is_repertoire (bool): whether the simulation is on a repertoire (person) or sequence/receptor level</p></li>
<li><p>paired: if the simulation should output paired data, this parameter should contain a list of a list of sim_item pairs referenced by name that should be combined; if paired data is not needed, then it should be False</p></li>
<li><p>sequence_type (str): either amino_acid or nucleotide</p></li>
<li><p>simulation_strategy (str): either RejectionSampling or Implanting, see the tutorials for more information on choosing one of these</p></li>
<li><p>keep_p_gen_dist (bool): if possible, whether to implement the importance sampling strategy, i.e., keep the distribution of generation probabilities of the sequences the same as provided by the model without any signals</p></li>
<li><p>p_gen_bin_count (int): if keep_p_gen_dist is true, how many bins to use to approximate the generation probability distribution</p></li>
<li><p>remove_seqs_with_signals (bool): if true, it explicitly controls the proportions of signals in sequences and removes any accidental occurrences</p></li>
<li><p>species (str): species that the sequences come from; used to select correct genes to export full length sequences; default is ‘human’</p></li>
<li><p>implanting_scaling_factor (int): determines in how many receptors to implant the signal in reach iteration; this is computed as number_of_receptors_needed_for_signal * implanting_scaling_factor; useful when using Implanting simulation strategy in combination with importance sampling, since the generation probability of some receptors with implanted signals might be very rare and those receptors might end up not being kept often with importance sampling; this parameter is only used when keep_p_gen_dist is set to True</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">simulations</span><span class="p">:</span>
<span class="w">  </span><span class="nt">sim1</span><span class="p">:</span>
<span class="w">    </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">    </span><span class="nt">paired</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">    </span><span class="nt">sequence_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">amino_acid</span>
<span class="w">    </span><span class="nt">simulation_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">RejectionSampling</span>
<span class="w">    </span><span class="nt">sim_items</span><span class="p">:</span>
<span class="w">      </span><span class="nt">sim_item1</span><span class="p">:</span><span class="w"> </span><span class="c1"># group of sequences with same simulation params</span>
<span class="w">        </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">          </span><span class="nt">chain</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta</span>
<span class="w">          </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">          </span><span class="nt">model_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">null</span>
<span class="w">          </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">        </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span>
<span class="w">        </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">         </span><span class="nt">signal1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
</pre></div>
</div>
</section>
<section id="simulation-config-item">
<h4>Simulation config item<a class="headerlink" href="#simulation-config-item" title="Link to this heading">¶</a></h4>
<p>When performing a simulation, one or more simulation config items can be specified. Config items define groups of repertoires or receptors
that have the same simulation parameters, such as signals, generative model, clonal frequencies, noise parameters.</p>
<p>Arguments:</p>
<ul>
<li><p>signals (dict): signals for the simulation item and the proportion of sequences in the repertoire that will have the given signal. For receptor-level simulation, the proportion will always be 1.</p></li>
<li><p>is_noise (bool): indicates whether the implanting should be regarded as noise; if it is True, the signals will be implanted as specified, but the repertoire/receptor in question will have negative class.</p></li>
<li><p>generative_model: parameters of the generative model, including its type, path to the model; currently supported models are OLGA and ExperimentalImport</p></li>
<li><p>false_positives_prob_in_receptors (float): when performing repertoire level simulation, what percentage of sequences should be false positives</p></li>
<li><p>false_negative_prob_in_receptors (float): when performing repertoire level simulation, what percentage of sequences should be false negatives</p></li>
<li><p>immune_events (dict): a set of key-value pairs that will be added to the metadata (same values for all data generated in one simulation sim_item) and can be later used as labels</p></li>
<li><p>default_clonal_frequency (dict): clonal frequency in Ligo is simulated through <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.zipf.html">scipy’s zeta distribution function for generating random numbers</a>, with parameters provided under default_clonal_frequency parameter. These parameters will be used to assign count values to sequences that do not contain any signals if they are required by the simulation. If clonal frequency shouldn’t be used, this parameter can be None</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">clonal_frequency</span><span class="p">:</span>
<span class="w">    </span><span class="nt">a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span><span class="w"> </span><span class="c1"># shape parameter of the distribution</span>
<span class="w">    </span><span class="nt">loc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span><span class="w"> </span><span class="c1"># 0 by default but can be used to shift the distribution</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>sequence_len_limits (dict): allows for filtering the generated sequences by length, needs to have parameters min and max specified; if not used, min/max should be -1</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">sequence_len_limits</span><span class="p">:</span>
<span class="w">    </span><span class="nt">min</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">4</span><span class="w"> </span><span class="c1"># keep sequences of length 4 and longer</span>
<span class="w">    </span><span class="nt">max</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">-1</span><span class="w"> </span><span class="c1"># no limit on the max length of the sequences</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">simulations</span><span class="p">:</span><span class="w"> </span><span class="c1"># definitions of simulations should be under key simulations in the definitions part of the specification</span>
<span class="w">    </span><span class="c1"># one simulation with multiple implanting objects, a part of definition section</span>
<span class="w">    </span><span class="nt">my_simulation</span><span class="p">:</span>
<span class="w">        </span><span class="nt">sim_item1</span><span class="p">:</span>
<span class="w">            </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="w">            </span><span class="nt">seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">null</span><span class="w"> </span><span class="c1"># don&#39;t use seed</span>
<span class="w">            </span><span class="nt">receptors_in_repertoire_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span>
<span class="w">            </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">                </span><span class="nt">chain</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta</span>
<span class="w">                </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">                </span><span class="nt">model_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">null</span>
<span class="w">                </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">            </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">                </span><span class="nt">my_signal</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.25</span>
<span class="w">                </span><span class="nt">my_signal2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.01</span>
<span class="w">                </span><span class="nt">my_signal__my_signal2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.02</span><span class="w"> </span><span class="c1"># my_signal and my_signal2 will co-occur in 2% of the receptors in all 10 repertoires</span>
<span class="w">        </span><span class="nt">sim_item2</span><span class="p">:</span>
<span class="w">            </span><span class="nt">number_of_examples</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">            </span><span class="nt">receptors_in_repertoire_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">150</span>
<span class="w">            </span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">                </span><span class="nt">chain</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta</span>
<span class="w">                </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
<span class="w">                </span><span class="nt">model_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">null</span>
<span class="w">                </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">            </span><span class="nt">signals</span><span class="p">:</span>
<span class="w">                </span><span class="nt">my_signal</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.75</span>
<span class="w">            </span><span class="nt">default_clonal_frequency</span><span class="p">:</span>
<span class="w">                </span><span class="nt">a</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="w">            </span><span class="nt">sequence_len_limits</span><span class="p">:</span>
<span class="w">                </span><span class="nt">min</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span>
</pre></div>
</div>
</section>
</section>
<section id="generative-models">
<h3>Generative models<a class="headerlink" href="#generative-models" title="Link to this heading">¶</a></h3>
<section id="experimentalimport">
<h4>ExperimentalImport<a class="headerlink" href="#experimentalimport" title="Link to this heading">¶</a></h4>
<p>Allows to import existing experimental data and perform annotations and simulations on top of them.</p>
<p>Arguments:</p>
<ul class="simple">
<li><p>import_format (str): see the list of supported formats under <a class="reference internal" href="#supported-dataset-formats"><span class="std std-ref">Supported dataset formats</span></a></p></li>
<li><p>tmp_import_path (str): where to store the imported files</p></li>
<li><p>import_params (dict): as defined under the import format selected in the first parameter; for details see <a class="reference internal" href="#supported-dataset-formats"><span class="std std-ref">Supported dataset formats</span></a></p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">    </span><span class="nt">import_format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AIRR</span>
<span class="w">    </span><span class="nt">tmp_import_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">./tmp/</span>
<span class="w">    </span><span class="nt">import_params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping AIRR: ligo</span>
<span class="w">            </span><span class="nt">junction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="w">            </span><span class="nt">junction_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="w">            </span><span class="nt">locus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chain</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ExperimentalImport</span>
</pre></div>
</div>
</section>
<section id="olga">
<h4>OLGA<a class="headerlink" href="#olga" title="Link to this heading">¶</a></h4>
<p>This is a wrapper for the OLGA package as described by Sethna et al. 2019 (OLGA package on PyPI or GitHub: <a class="reference external" href="https://github.com/statbiophys/OLGA">https://github.com/statbiophys/OLGA</a>).</p>
<p>Reference:</p>
<p>Zachary Sethna, Yuval Elhanati, Curtis G Callan, Jr, Aleksandra M Walczak, Thierry Mora, OLGA: fast computation of generation probabilities
of B- and T-cell receptor amino acid sequences and motifs, Bioinformatics, Volume 35, Issue 17, 1 September 2019, Pages 2974–2981,
<a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz035">https://doi.org/10.1093/bioinformatics/btz035</a></p>
<p>Note:</p>
<ul class="simple">
<li><p>OLGA generates sequences that correspond to IMGT junction and are used for matching as such. See the <a class="reference external" href="https://github.com/statbiophys/OLGA">https://github.com/statbiophys/OLGA</a> for more details.</p></li>
<li><p>Gene names are as provided in OLGA (either in default models or in the user-specified model files). For simulation, one should use gene names in the same format.</p></li>
</ul>
<p>Arguments:</p>
<ul class="simple">
<li><p>model_path (str): if not default model, this parameter should point to a folder where the four OLGA/IGOR format files are stored (could also be inferred from some experimental data)</p></li>
<li><p>default_model_name (str): if not using custom models, one of the OLGA default models could be specified here; the value should be the same as it would be passed to command line in OLGA: e.g., humanTRB, human IGH</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">generative_model</span><span class="p">:</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">    </span><span class="nt">model_path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">None</span>
<span class="w">    </span><span class="nt">default_model_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">humanTRB</span>
</pre></div>
</div>
</section>
</section>
<section id="supported-dataset-formats">
<h3>Supported dataset formats<a class="headerlink" href="#supported-dataset-formats" title="Link to this heading">¶</a></h3>
<p>These dataset formats can be used in combination with <a class="reference internal" href="specs/definitions/gen_models.html#experimentalimport"><span class="std std-ref">ExperimentalImport</span></a> generative model.</p>
<section id="airr">
<h4>AIRR<a class="headerlink" href="#airr" title="Link to this heading">¶</a></h4>
<p>Imports data in AIRR format. The AIRR .tsv format is explained here: <a class="reference external" href="https://docs.airr-community.org/en/stable/datarep/format.html">https://docs.airr-community.org/en/stable/datarep/format.html</a>
And the AIRR rearrangement schema can be found here: <a class="reference external" href="https://docs.airr-community.org/en/stable/datarep/rearrangements.html">https://docs.airr-community.org/en/stable/datarep/rearrangements.html</a></p>
<p>Arguments:</p>
<ul>
<li><p>path (str): Either the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): Should be set to False, as we import a SequenceDataset.</p></li>
<li><p>import_productive (bool): Whether productive sequences (with value ‘T’ in column productive) should be included in the imported sequences. By default, import_productive is True.</p></li>
<li><p>import_with_stop_codon (bool): Whether sequences with stop codons (with value ‘T’ in column stop_codon) should be included in the imported sequences. This only applies if column stop_codon is present. By default, import_with_stop_codon is False.</p></li>
<li><p>import_out_of_frame (bool): Whether out of frame sequences (with value ‘F’ in column vj_in_frame) should be included in the imported sequences. This only applies if column vj_in_frame is present. By default, import_out_of_frame is False.</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as AIRR uses the IMGT junction. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): A mapping from AIRR column names to immuneML’s internal data representation. A custom column mapping can be specified here if necessary (for example; adding additional data fields if they are present in the AIRR file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’].. For AIRR, this is by default set to:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">junction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="nt">junction_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="nt">locus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chain</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For AIRR format, there is no default column_mapping_synonyms.</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are AIRR column names and values are the names that are internally used in immuneML as metadata fields. These metadata fields can be used as prediction labels for Sequence- and ReceptorDatasets. This parameter can also be used to specify sequence-level metadata columns for RepertoireDatasets, which can be used by reports. To set prediction label metadata for RepertoireDatasets, see metadata_file instead. For AIRR format, there is no default metadata_column_mapping.</p></li>
<li><p>separator (str): Column separator, for AIRR this is by default “t”.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_airr_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AIRR</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="w">        </span><span class="nt">import_productive</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to include productive sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_with_stop_codon</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include sequences with stop codon in the dataset</span>
<span class="w">        </span><span class="nt">import_out_of_frame</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include out of frame sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even if the `sequences` column is empty (provided that other fields are as specified here)</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove all sequences with empty `sequence_aa` column</span>
<span class="w">        </span><span class="c1"># Optional fields with AIRR-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping AIRR: immuneML</span>
<span class="w">            </span><span class="nt">junction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="w">            </span><span class="nt">junction_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="w">            </span><span class="nt">locus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chain</span>
</pre></div>
</div>
</section>
<section id="generic">
<h4>Generic<a class="headerlink" href="#generic" title="Link to this heading">¶</a></h4>
<p>Imports data from any tabular file into a Repertoire-, Sequence- or ReceptorDataset. RepertoireDatasets should be
used when making predictions per repertoire, such as predicting a disease state. SequenceDatasets or ReceptorDatasets
should be used when predicting values for unpaired (single-chain) and paired immune receptors respectively,
like antigen specificity.</p>
<p>This importer works similarly to other importers, but has no predefined default values for which fields are imported,
and can therefore be tailored to import data from various different tabular files with headers.</p>
<p>For ReceptorDatasets: this importer assumes the two receptor sequences appear on different lines in the file, and can
be paired together by a common sequence identifier. If you instead want to import a ReceptorDataset from a tabular
file that contains both receptor chains on one line, see <a class="reference internal" href="specs/definitions/datasets.html#singlelinereceptor"><span class="std std-ref">SingleLineReceptor</span></a> import</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): For RepertoireDatasets, this is the path to a directory with files to import. For Sequence- or ReceptorDatasets this path may either be the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset or ReceptorDataset. By default, is_repertoire is set to True.</p></li>
<li><p>metadata_file (str): Required for RepertoireDatasets. This parameter specifies the path to the metadata file. This is a csv file with columns filename, subject_id and arbitrary other columns which can be used as labels in instructions. For setting Sequence- or ReceptorDataset metadata, metadata_file is ignored, see metadata_column_mapping instead.</p></li>
<li><p>paired (str): Required for Sequence- or ReceptorDatasets. This parameter determines whether to import a SequenceDataset (paired = False) or a ReceptorDataset (paired = True). In a ReceptorDataset, two sequences with chain types specified by receptor_chains are paired together based on a common identifier. This identifier should be mapped to the immuneML field ‘sequence_identifiers’ using the column_mapping.</p></li>
<li><p>receptor_chains (str): Required for ReceptorDatasets. Determines which pair of chains to import for each Receptor. Valid values are <cite>TRA_TRB</cite>, <cite>TRG_TRD</cite>, <cite>IGH_IGL</cite>, <cite>IGH_IGK</cite>.</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means immuneML assumes the IMGT junction (including leading C and trailing Y/F amino acids) is used in the input file, and the first and last amino acids will be removed from the sequences to retrieve the IMGT CDR3 sequence. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): Required for all datasets. A mapping where the keys are the column names in the input file, and the values correspond to the names used in immuneML’s internal data representation. Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’].. At least sequences (nucleotide) or sequence_aas (amino acids) must be specified, but all other fields are optional. A column mapping can look for example like this:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">file_column_amino_acids</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="nt">file_column_v_genes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="nt">file_column_j_genes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="nt">file_column_frequencies</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For Generic import, there is no default column_mapping_synonyms.</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are file column names and values are the names that are internally used in immuneML as metadata fields. These metadata fields can be used as prediction labels for Sequence- and ReceptorDatasets. This parameter can also be used to specify sequence-level metadata columns for RepertoireDatasets, which can be used by reports. To set prediction label metadata for RepertoireDatasets, see metadata_file instead. There is no default metadata_column_mapping.</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">file_column_antigen_specificity</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">antigen_specificity</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>columns_to_load (list): Optional; specifies which columns to load from the input file. This may be useful if the input files contain many unused columns. If no value is specified, all columns are loaded.</p></li>
<li><p>separator (str): Required parameter. Column separator, for example “t” or “,”. The default value is “t”</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_generic_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Generic</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span><span class="w"> </span><span class="c1"># metadata file for RepertoireDataset</span>
<span class="w">        </span><span class="nt">paired</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to import SequenceDataset (False) or ReceptorDataset (True) when is_repertoire = False</span>
<span class="w">        </span><span class="nt">receptor_chains</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRA_TRB</span><span class="w"> </span><span class="c1"># what chain pair to import for a ReceptorDataset</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># filter out sequences if they don&#39;t have sequence_aa set</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping file: immuneML</span>
<span class="w">            </span><span class="nt">file_column_amino_acids</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aas</span>
<span class="w">            </span><span class="nt">file_column_v_genes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="w">            </span><span class="nt">file_column_j_genes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="w">            </span><span class="nt">file_column_frequencies</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># metadata column mapping file: immuneML</span>
<span class="w">            </span><span class="nt">file_column_antigen_specificity</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">antigen_specificity</span>
<span class="w">        </span><span class="nt">columns_to_load</span><span class="p">:</span><span class="w">  </span><span class="c1"># which subset of columns to load from the file</span>
<span class="w">            </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">file_column_amino_acids</span>
<span class="w">            </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">file_column_v_genes</span>
<span class="w">            </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">file_column_j_genes</span>
<span class="w">            </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">file_column_frequencies</span>
<span class="w">            </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">file_column_antigen_specificity</span>
</pre></div>
</div>
</section>
<section id="igor">
<h4>IGoR<a class="headerlink" href="#igor" title="Link to this heading">¶</a></h4>
<p>Imports data generated by <a class="reference external" href="https://github.com/qmarcou/IGoR">IGoR</a> simulations into a Repertoire-, or SequenceDataset.
RepertoireDatasets should be used when making predictions per repertoire, such as predicting a disease state.
SequenceDatasets should be used when predicting values for unpaired (single-chain) immune receptors, like
antigen specificity.</p>
<p>Note that you should run IGoR with the –CDR3 option specified, this tool imports the generated CDR3 files.
Sequences with missing anchors are not imported, meaning only sequences with value ‘1’ in the anchors_found column are imported.
Nucleotide sequences are automatically translated to amino acid sequences.</p>
<p>Reference: Quentin Marcou, Thierry Mora, Aleksandra M. Walczak
‘High-throughput immune repertoire analysis with IGoR’. Nature Communications, (2018)
<a class="reference external" href="https://doi.org/10.1038/s41467-018-02832-w">doi.org/10.1038/s41467-018-02832-w</a>.</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): For RepertoireDatasets, this is the path to a directory with IGoR files to import. For Sequence- or ReceptorDatasets this path may either be the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset. By default, is_repertoire is set to True.</p></li>
<li><p>metadata_file (str): Required for RepertoireDatasets. This parameter specifies the path to the metadata file. This is a csv file with columns filename, subject_id and arbitrary other columns which can be used as labels in instructions. Only the IGoR files included under the column ‘filename’ are imported into the RepertoireDataset. For setting SequenceDataset metadata, metadata_file is ignored, see metadata_column_mapping instead.</p></li>
<li><p>import_with_stop_codon (bool): Whether sequences with stop codons should be included in the imported sequences. By default, import_with_stop_codon is False.</p></li>
<li><p>import_out_of_frame (bool): Whether out of frame sequences (with value ‘0’ in column is_inframe) should be included in the imported sequences. By default, import_out_of_frame is False.</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as IGoR uses the IMGT junction. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): A mapping from IGoR column names to immuneML’s internal data representation. A custom column mapping can be specified here if necessary (for example; adding additional data fields if they are present in the IGoR file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’].. For IGoR, this is by default set to:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">nt_CDR3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequences</span>
<span class="nt">seq_index</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_identifiers</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For IGoR format, there is no default column_mapping_synonyms.</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are IGoR column names and values are the names that are internally used in immuneML as metadata fields. These metadata fields can be used as prediction labels for Sequence- and ReceptorDatasets. This parameter can also be used to specify sequence-level metadata columns for RepertoireDatasets, which can be used by reports. To set prediction label metadata for RepertoireDatasets, see metadata_file instead. For IGoR format, there is no default metadata_column_mapping.</p></li>
<li><p>separator (str): Column separator, for IGoR this is by default “,”.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_igor_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IGoR</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset (True) or a SequenceDataset (False)</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span><span class="w"> </span><span class="c1"># metadata file for RepertoireDataset</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># metadata column mapping IGoR: immuneML for SequenceDataset</span>
<span class="w">            </span><span class="nt">igor_column_name1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label1</span>
<span class="w">            </span><span class="nt">igor_column_name2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label2</span>
<span class="w">        </span><span class="nt">import_with_stop_codon</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include sequences with stop codon in the dataset</span>
<span class="w">        </span><span class="nt">import_out_of_frame</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include out of frame sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="c1"># Optional fields with IGoR-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;,&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping IGoR: immuneML</span>
<span class="w">            </span><span class="nt">nt_CDR3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequences</span>
<span class="w">            </span><span class="nt">seq_index</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_identifiers</span>
</pre></div>
</div>
</section>
<section id="ireceptor">
<h4>IReceptor<a class="headerlink" href="#ireceptor" title="Link to this heading">¶</a></h4>
<p>Imports AIRR datasets retrieved through the <a class="reference external" href="https://gateway.ireceptor.org/home">iReceptor Gateway</a> into a Repertoire-, Sequence- or ReceptorDataset.
The differences between this importer and the <a class="reference internal" href="specs/definitions/datasets.html#airr"><span class="std std-ref">AIRR</span></a> importer are:</p>
<ul class="simple">
<li><p>This importer takes in a list of .zip files, which must contain one or more AIRR tsv files, and for each AIRR file, a corresponding metadata json file must be present.</p></li>
<li><p>This importer does not require a metadata csv file for RepertoireDataset import, it is generated automatically from the metadata json files.</p></li>
</ul>
<p>RepertoireDatasets should be used when making predictions per repertoire, such as predicting a disease state.
SequenceDatasets or ReceptorDatasets should be used when predicting values for unpaired (single-chain) and paired
immune receptors respectively, like antigen specificity.</p>
<p>AIRR rearrangement schema can be found here: <a class="reference external" href="https://docs.airr-community.org/en/stable/datarep/rearrangements.html">https://docs.airr-community.org/en/stable/datarep/rearrangements.html</a></p>
<p>When importing a ReceptorDataset, the AIRR field cell_id is used to determine the chain pairs.</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): This is the path to a directory <strong>with .zip files</strong> retrieved from the iReceptor Gateway. These .zip files should include AIRR files (with .tsv extension) and corresponding metadata.json files with matching names (e.g., for my_dataset.tsv the corresponding metadata file is called my_dataset-metadata.json). The zip files must use the .zip extension.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset or ReceptorDataset. By default, is_repertoire is set to True.</p></li>
<li><p>paired (str): Required for Sequence- or ReceptorDatasets. This parameter determines whether to import a SequenceDataset (paired = False) or a ReceptorDataset (paired = True). In a ReceptorDataset, two sequences with chain types specified by receptor_chains are paired together based on the identifier given in the AIRR column named ‘cell_id’.</p></li>
<li><p>receptor_chains (str): Required for ReceptorDatasets. Determines which pair of chains to import for each Receptor. Valid values are <cite>TRA_TRB</cite>, <cite>TRG_TRD</cite>, <cite>IGH_IGL</cite>, <cite>IGH_IGK</cite>. If receptor_chains is not provided, the chain pair is automatically detected (only one chain pair type allowed per repertoire).</p></li>
<li><p>import_productive (bool): Whether productive sequences (with value ‘T’ in column productive) should be included in the imported sequences. By default, import_productive is True.</p></li>
<li><p>import_with_stop_codon (bool): Whether sequences with stop codons (with value ‘T’ in column stop_codon) should be included in the imported sequences. This only applies if column stop_codon is present. By default, import_with_stop_codon is False.</p></li>
<li><p>import_out_of_frame (bool): Whether out of frame sequences (with value ‘F’ in column vj_in_frame) should be included in the imported sequences. This only applies if column vj_in_frame is present. By default, import_out_of_frame is False.</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as AIRR uses the IMGT junction. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): A mapping from AIRR column names to immuneML’s internal data representation. For AIRR, this is by default set to the values shown in YAML below.  A custom column mapping can be specified here if necessary (for example; adding additional data fields if they are present in the AIRR file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’]. A custom column mapping can be specified here if necessary (for example; adding additional data fields if they are present in the AIRR file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’]..</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">junction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequences</span>
<span class="nt">junction_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aas</span>
<span class="nt">v_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_alleles</span>
<span class="nt">j_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_alleles</span>
<span class="nt">locus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chains</span>
<span class="nt">duplicate_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">counts</span>
<span class="nt">sequence_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_identifiers</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For AIRR format, there is no default column_mapping_synonyms.</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are AIRR column names and values are the names that are internally used in immuneML as metadata fields. These metadata fields can be used as prediction labels for Sequence- and ReceptorDatasets. For AIRR format, there is no default metadata_column_mapping. When importing a RepertoireDataset, the metadata is automatically extracted from the metadata json files.</p></li>
<li><p>separator (str): Column separator, for AIRR this is by default “t”.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_airr_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IReceptor</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/zipfiles/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># metadata column mapping AIRR: immuneML for Sequence- or ReceptorDatasetDataset</span>
<span class="w">            </span><span class="nt">airr_column_name1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label1</span>
<span class="w">            </span><span class="nt">airr_column_name2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label2</span>
<span class="w">        </span><span class="nt">import_productive</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to include productive sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_with_stop_codon</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include sequences with stop codon in the dataset</span>
<span class="w">        </span><span class="nt">import_out_of_frame</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include out of frame sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even if the `sequences` column is empty (provided that other fields are as specified here)</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove all sequences with empty `sequence_aas` column</span>
<span class="w">        </span><span class="c1"># Optional fields with AIRR-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping AIRR: immuneML</span>
<span class="w">            </span><span class="nt">junction</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequences</span>
<span class="w">            </span><span class="nt">junction_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aas</span>
<span class="w">            </span><span class="nt">v_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_alleles</span>
<span class="w">            </span><span class="nt">j_call</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_alleles</span>
<span class="w">            </span><span class="nt">locus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chains</span>
<span class="w">            </span><span class="nt">duplicate_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">counts</span>
<span class="w">            </span><span class="nt">sequence_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_identifiers</span>
</pre></div>
</div>
</section>
<section id="immuneml">
<h4>ImmuneML<a class="headerlink" href="#immuneml" title="Link to this heading">¶</a></h4>
<p>Imports the dataset from the files previously exported by immuneML. It closely resembles the AIRR format but relies on binary
representations and is optimized for faster read-in at runtime.</p>
<p>ImmuneMLImport can import any kind of dataset (RepertoireDataset, SequenceDataset, ReceptorDataset).</p>
<p>This format includes:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>a dataset file in yaml format with iml_dataset extension with parameters:</p>
<ul class="simple">
<li><p>name,</p></li>
<li><p>identifier,</p></li>
<li><p>metadata_file (for repertoire datasets),</p></li>
<li><p>metadata_fields (for repertoire datasets),</p></li>
<li><p>repertoire_ids (for repertoire datasets)</p></li>
<li><p>element_ids (for receptor and sequence datasets),</p></li>
<li><p>labels,</p></li>
</ul>
</li>
<li><p>a csv metadata file (only for repertoire datasets, should be in the same folder as the iml_dataset file),</p></li>
<li><p>data files for different types of data. For repertoire datasets, data files include one binary numpy file per repertoire with sequences and associated information and one metadata yaml file per repertoire with details such as repertoire identifier, disease status, subject id and other similar available information. For sequence and receptor datasets, sequences or receptors respectively, are stored in batches in binary numpy files.</p></li>
</ol>
</div></blockquote>
<p>Arguments:</p>
<ul class="simple">
<li><p>path (str): The path to the previously created dataset file. This file should have an ‘.iml_dataset’ extension. If the path has not been specified, immuneML attempts to load the dataset from a specified metadata file (only for RepertoireDatasets).</p></li>
<li><p>metadata_file (str): An optional metadata file for a RepertoireDataset. If specified, the RepertoireDataset metadata will be updated to the newly specified metadata without otherwise changing the Repertoire objects</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ImmuneML</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/dataset.iml_dataset</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span>
</pre></div>
</div>
</section>
<section id="immunoseqrearrangement">
<h4>ImmunoSEQRearrangement<a class="headerlink" href="#immunoseqrearrangement" title="Link to this heading">¶</a></h4>
<p>Imports data from Adaptive Biotechnologies immunoSEQ Analyzer rearrangement-level .tsv files into a
Repertoire-, or SequenceDataset.
RepertoireDatasets should be used when making predictions per repertoire, such as predicting a disease state.
SequenceDatasets should be used when predicting values for unpaired (single-chain) immune receptors, like
antigen specificity.</p>
<p>The format of the files imported by this importer is described here:
<a class="reference external" href="https://www.adaptivebiotech.com/wp-content/uploads/2019/07/MRK-00342_immunoSEQ_TechNote_DataExport_WEB_REV.pdf">https://www.adaptivebiotech.com/wp-content/uploads/2019/07/MRK-00342_immunoSEQ_TechNote_DataExport_WEB_REV.pdf</a>
Alternatively, to import sample-level .tsv files, see <a class="reference internal" href="specs/definitions/datasets.html#immunoseqsample"><span class="std std-ref">ImmunoSEQSample</span></a> import</p>
<p>The only difference between these two importers is which columns they load from the .tsv files.</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): For RepertoireDatasets, this is the path to a directory with files to import. For Sequence- or ReceptorDatasets this path may either be the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset. By default, is_repertoire is set to True.</p></li>
<li><p>metadata_file (str): Required for RepertoireDatasets. This parameter specifies the path to the metadata file. This is a csv file with columns filename, subject_id and arbitrary other columns which can be used as labels in instructions. Only the files included under the column ‘filename’ are imported into the RepertoireDataset. For setting SequenceDataset metadata, metadata_file is ignored, see metadata_column_mapping instead.</p></li>
<li><p>import_productive (bool): Whether productive sequences (with value ‘In’ in column frame_type) should be included in the imported sequences. By default, import_productive is True.</p></li>
<li><p>import_with_stop_codon (bool): Whether sequences with stop codons (with value ‘Stop’ in column frame_type) should be included in the imported sequences. By default, import_with_stop_codon is False.</p></li>
<li><p>import_out_of_frame (bool): Whether out of frame sequences (with value ‘Out’ in column frame_type) should be included in the imported sequences. By default, import_out_of_frame is False.</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as immunoSEQ files use the IMGT junction. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): A mapping from immunoSEQ column names to immuneML’s internal data representation. For immunoSEQ rearrangement-level files, this is by default set the values shown below in YAML format.         A custom column mapping can be specified here if necessary (for example: adding additional data fields if they are present in the file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’]..</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">rearrangement</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="nt">amino_acid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="nt">v_resolved</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="nt">j_resolved</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="nt">templates</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="nt">locus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chain</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For immunoSEQ rearrangement-level files, this is by default set to:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">v_resolved</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_alleles</span>
<span class="nt">j_resolved</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_alleles</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>columns_to_load (list): Specifies which subset of columns must be loaded from the file. By default, this is: [rearrangement, v_family, v_gene, v_allele, j_family, j_gene, j_allele, amino_acid, templates, frame_type, locus]</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are immunoSEQ column names and values are the names that are internally used in immuneML as metadata fields. These metadata fields can be used as prediction labels for Sequence- and ReceptorDatasets. This parameter can also be used to specify sequence-level metadata columns for RepertoireDatasets, which can be used by reports. To set prediction label metadata for RepertoireDatasets, see metadata_file instead. For immunoSEQ rearrangement .tsv files, there is no default metadata_column_mapping.</p></li>
<li><p>separator (str): Column separator, for ImmunoSEQ files this is by default “t”.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter will typically be False (import only non-empty amino acid sequences)</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_immunoseq_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ImmunoSEQRearrangement</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset (True) or a SequenceDataset (False)</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span><span class="w"> </span><span class="c1"># metadata file for RepertoireDataset</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># metadata column mapping ImmunoSEQ: immuneML for SequenceDataset</span>
<span class="w">            </span><span class="nt">immunoseq_column_name1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label1</span>
<span class="w">            </span><span class="nt">immunoseq_column_name2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label2</span>
<span class="w">        </span><span class="nt">import_productive</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to include productive sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_with_stop_codon</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include sequences with stop codon in the dataset</span>
<span class="w">        </span><span class="nt">import_out_of_frame</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include out of frame sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># filter out sequences if they don&#39;t have sequence_aa set</span>
<span class="w">        </span><span class="c1"># Optional fields with ImmunoSEQ rearrangement-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">columns_to_load</span><span class="p">:</span><span class="w"> </span><span class="c1"># subset of columns to load</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">rearrangement</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_family</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_gene</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_resolved</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_family</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_gene</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_resolved</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">amino_acid</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">templates</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">frame_type</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">locus</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping immunoSEQ: immuneML</span>
<span class="w">            </span><span class="nt">rearrangement</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="w">            </span><span class="nt">amino_acid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="w">            </span><span class="nt">v_resolved</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="w">            </span><span class="nt">j_resolved</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="w">            </span><span class="nt">templates</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="w">            </span><span class="nt">locus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chain</span>
</pre></div>
</div>
</section>
<section id="immunoseqsample">
<h4>ImmunoSEQSample<a class="headerlink" href="#immunoseqsample" title="Link to this heading">¶</a></h4>
<p>Imports data from Adaptive Biotechnologies immunoSEQ Analyzer sample-level .tsv files into a
Repertoire-, or SequenceDataset.
RepertoireDatasets should be used when making predictions per repertoire, such as predicting a disease state.
SequenceDatasets should be used when predicting values for unpaired (single-chain) immune receptors, like
antigen specificity.</p>
<p>The format of the files imported by this importer is described here in section 3.4.13
<a class="reference external" href="https://clients.adaptivebiotech.com/assets/downloads/immunoSEQ_AnalyzerManual.pdf">https://clients.adaptivebiotech.com/assets/downloads/immunoSEQ_AnalyzerManual.pdf</a>
Alternatively, to import rearrangement-level .tsv files, see <a class="reference internal" href="specs/definitions/datasets.html#immunoseqrearrangement"><span class="std std-ref">ImmunoSEQRearrangement</span></a> import.
The only difference between these two importers is which columns they load from the .tsv files.</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): For RepertoireDatasets, this is the path to a directory with files to import. For Sequence- or ReceptorDatasets this path may either be the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset. By default, is_repertoire is set to True.</p></li>
<li><p>metadata_file (str): Required for RepertoireDatasets. This parameter specifies the path to the metadata file. This is a csv file with columns filename, subject_id and arbitrary other columns which can be used as labels in instructions. Only the files included under the column ‘filename’ are imported into the RepertoireDataset. For setting SequenceDataset metadata, metadata_file is ignored, see metadata_column_mapping instead.</p></li>
<li><p>import_productive (bool): Whether productive sequences (with value ‘In’ in column frame_type) should be included in the imported sequences. By default, import_productive is True.</p></li>
<li><p>import_with_stop_codon (bool): Whether sequences with stop codons (with value ‘Stop’ in column frame_type) should be included in the imported sequences. By default, import_with_stop_codon is False.</p></li>
<li><p>import_out_of_frame (bool): Whether out of frame sequences (with value ‘Out’ in column frame_type) should be included in the imported sequences. By default, import_out_of_frame is False.</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as immunoSEQ files use the IMGT junction. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): A mapping from immunoSEQ column names to immuneML’s internal data representation. For immunoSEQ sample-level files, this is by default set to the values shown bellow in YAML format.         A custom column mapping can be specified here if necessary (for example; adding additional data fields if they are present in the file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’]..</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">nucleotide</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="nt">aminoAcid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="nt">vGeneName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="nt">jGeneName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="nt">sequenceStatus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">frame_type</span>
<span class="nt">vFamilyName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_family</span>
<span class="nt">jFamilyName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_family</span>
<span class="nt">vGeneAllele</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_allele</span>
<span class="nt">jGeneAllele</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_allele</span>
<span class="nt">count (templates/reads)</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For immunoSEQ sample .tsv files, there is no default column_mapping_synonyms.</p></li>
<li><p>columns_to_load (list): Specifies which subset of columns must be loaded from the file. By default, this is: [nucleotide, aminoAcid, count (templates/reads), vFamilyName, vGeneName, vGeneAllele, jFamilyName, jGeneName, jGeneAllele, sequenceStatus]; these are the columns from the original file that will be imported</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are immunoSEQ column names and values are the names that are internally used in immuneML as metadata fields. These metadata fields can be used as prediction labels for Sequence- and ReceptorDatasets. This parameter can also be used to specify sequence-level metadata columns for RepertoireDatasets, which can be used by reports. To set prediction label metadata for RepertoireDatasets, see metadata_file instead. For immunoSEQ sample .tsv files, there is no default metadata_column_mapping.</p></li>
<li><p>separator (str): Column separator, for ImmunoSEQ files this is by default “t”.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_immunoseq_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ImmunoSEQSample</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset (True) or a SequenceDataset (False)</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span><span class="w"> </span><span class="c1"># metadata file for RepertoireDataset</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># metadata column mapping ImmunoSEQ: immuneML for SequenceDataset</span>
<span class="w">            </span><span class="nt">immunoseq_column_name1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label1</span>
<span class="w">            </span><span class="nt">immunoseq_column_name2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label2</span>
<span class="w">        </span><span class="nt">import_productive</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to include productive sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_with_stop_codon</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include sequences with stop codon in the dataset</span>
<span class="w">        </span><span class="nt">import_out_of_frame</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to include out of frame sequences in the dataset</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># filter out sequences if they don&#39;t have sequence_aa set</span>
<span class="w">        </span><span class="c1"># Optional fields with ImmunoSEQ sample-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">columns_to_load</span><span class="p">:</span><span class="w"> </span><span class="c1"># subset of columns to load</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">nucleotide</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">aminoAcid</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">count (templates/reads)</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">vFamilyName</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">vGeneName</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">vGeneAllele</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">jFamilyName</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">jGeneName</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">jGeneAllele</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequenceStatus</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping immunoSEQ: immuneML</span>
<span class="w">            </span><span class="nt">nucleotide</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="w">            </span><span class="nt">aminoAcid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="w">            </span><span class="nt">vGeneName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="w">            </span><span class="nt">jGeneName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="w">            </span><span class="nt">sequenceStatus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">frame_type</span>
<span class="w">            </span><span class="nt">vFamilyName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_family</span>
<span class="w">            </span><span class="nt">jFamilyName</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_family</span>
<span class="w">            </span><span class="nt">vGeneAllele</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_allele</span>
<span class="w">            </span><span class="nt">jGeneAllele</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_allele</span>
<span class="w">            </span><span class="nt">count (templates/reads)</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
</pre></div>
</div>
</section>
<section id="mixcr">
<h4>MiXCR<a class="headerlink" href="#mixcr" title="Link to this heading">¶</a></h4>
<p>Imports data in MiXCR format into a Repertoire-, or SequenceDataset.
RepertoireDatasets should be used when making predictions per repertoire, such as predicting a disease state.
SequenceDatasets should be used when predicting values for unpaired (single-chain) immune receptors, like
antigen specificity.</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): For RepertoireDatasets, this is the path to a directory with MiXCR files to import. For Sequence- or ReceptorDatasets this path may either be the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset. By default, is_repertoire is set to True.</p></li>
<li><p>metadata_file (str): Required for RepertoireDatasets. This parameter specifies the path to the metadata file. This is a csv file with columns filename, subject_id and arbitrary other columns which can be used as labels in instructions. Only the MiXCR files included under the column ‘filename’ are imported into the RepertoireDataset. For setting SequenceDataset metadata, metadata_file is ignored, see metadata_column_mapping instead.</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence, such as ‘_’, are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as MiXCR uses IMGT junction as CDR3. Alternatively to importing the CDR3 sequence, other region types can be specified here as well. Valid values are <cite>IMGT_CDR3</cite>, <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>.</p></li>
<li><p>column_mapping (dict): A mapping from MiXCR column names to immuneML’s internal data representation. The columns that specify the sequences to import are handled by the region_type parameter. A custom column mapping can be specified here if necessary (for example; adding additional data fields if they are present in the MiXCR file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’].. For MiXCR, this is by default set to:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">cloneCount</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="nt">allVHitsWithScore</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="nt">allJHitsWithScore</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For MiXCR format, there is no default column_mapping_synonyms.</p></li>
<li><p>columns_to_load (list): Specifies which subset of columns must be loaded from the MiXCR file. By default, this is: [cloneCount, allVHitsWithScore, allJHitsWithScore, aaSeqCDR3, nSeqCDR3]</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are MiXCR column names and values are the names that are internally used in immuneML as metadata fields. These metadata fields can be used as prediction labels for Sequence- and ReceptorDatasets. This parameter can also be used to specify sequence-level metadata columns for RepertoireDatasets, which can be used by reports. To set prediction label metadata for RepertoireDatasets, see metadata_file instead. For MiXCR format, there is no default metadata_column_mapping.</p></li>
<li><p>separator (str): Column separator, for MiXCR this is by default “t”.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_mixcr_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">MiXCR</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset (True) or a SequenceDataset (False)</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span><span class="w"> </span><span class="c1"># metadata file for RepertoireDataset</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># metadata column mapping MiXCR: immuneML for SequenceDataset</span>
<span class="w">            </span><span class="nt">mixcrColumnName1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label1</span>
<span class="w">            </span><span class="nt">mixcrColumnName2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label2</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># filter out sequences if they don&#39;t have sequence_aa set</span>
<span class="w">        </span><span class="c1"># Optional fields with MiXCR-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">columns_to_load</span><span class="p">:</span><span class="w"> </span><span class="c1"># subset of columns to load, sequence columns are handled by region_type parameter</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cloneCount</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">allVHitsWithScore</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">allJHitsWithScore</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">aaSeqCDR3</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">nSeqCDR3</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping MiXCR: immuneML</span>
<span class="w">            </span><span class="nt">cloneCount</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="w">            </span><span class="nt">allVHitsWithScore</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="w">            </span><span class="nt">allJHitsWithScore</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
</pre></div>
</div>
</section>
<section id="id3">
<h4>OLGA<a class="headerlink" href="#id3" title="Link to this heading">¶</a></h4>
<p>Imports data generated by <a class="reference external" href="https://github.com/statbiophys/OLGA">OLGA</a> simulations into a Repertoire-, or SequenceDataset. Assumes that the columns in each
file correspond to: nucleotide sequences, amino acid sequences, v genes, j genes</p>
<p>Reference: Sethna, Zachary et al.
‘High-throughput immune repertoire analysis with IGoR’. Bioinformatics, (2019)
<a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz035">doi.org/10.1093/bioinformatics/btz035</a>.</p>
<p>Arguments:</p>
<ul class="simple">
<li><p>path (str): For RepertoireDatasets, this is the path to a directory with OLGA files to import. For Sequence- or ReceptorDatasets this path may either be the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset. By default, is_repertoire is set to True.</p></li>
<li><p>metadata_file (str): Required for RepertoireDatasets. This parameter specifies the path to the metadata file. This is a csv file with columns filename, subject_id and arbitrary other columns which can be used as labels in instructions. Only the OLGA files included under the column ‘filename’ are imported into the RepertoireDataset. SequenceDataset metadata is currently not supported.</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as OLGA uses the IMGT junction. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>separator (str): Column separator, for OLGA this is by default “t”.</p></li>
<li><p>column_mapping (dict): defines which columns to import from olga format: keys are the number of the columns and values are the names of the columns to be mapped to</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_olga_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">OLGA</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset (True) or a SequenceDataset (False)</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span><span class="w"> </span><span class="c1"># metadata file for RepertoireDataset</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># filter out sequences if they don&#39;t have sequence_aa set</span>
<span class="w">        </span><span class="c1"># Optional fields with OLGA-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">columns_to_load</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">0</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">1</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">2</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">3</span><span class="p p-Indicator">]</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span>
<span class="w">            </span><span class="nt">0</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequences</span>
<span class="w">            </span><span class="nt">1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aas</span>
<span class="w">            </span><span class="nt">2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_genes</span>
<span class="w">            </span><span class="nt">3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_genes</span>
</pre></div>
</div>
</section>
<section id="singlelinereceptor">
<h4>SingleLineReceptor<a class="headerlink" href="#singlelinereceptor" title="Link to this heading">¶</a></h4>
<p>Imports data from a tabular file (where each line contains a pair of immune receptor sequences) into a ReceptorDataset.
If you instead want to import a ReceptorDataset from a tabular file that contains one receptor sequence per line,
see <a class="reference internal" href="specs/definitions/datasets.html#generic"><span class="std std-ref">Generic</span></a> import.</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): Required parameter. This is the path to a directory with files to import.</p></li>
<li><p>receptor_chains (str): Required parameter. Determines which pair of chains to import for each Receptor. Valid values for receptor_chains are: <cite>TRA_TRB</cite>, <cite>TRG_TRD</cite>, <cite>IGH_IGL</cite>, <cite>IGH_IGK</cite>.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. When IMGT_CDR3 is specified, immuneML assumes the IMGT junction (including leading C and trailing Y/F amino acids) is used in the input file, and the first and last amino acids will be removed from the sequences to retrieve the IMGT CDR3 sequence. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): A mapping where the keys are the column names in the input file, and the values must be mapped to the following fields: &lt;chain&gt;_amino_acid_sequence, &lt;chain&gt;_nucleotide_sequence, &lt;chain&gt;_v_gene, &lt;chain&gt;_j_gene, identifier, epitope. The possible names that can be filled in for &lt;chain&gt; are: <cite>ALPHA</cite>, <cite>BETA</cite>, <cite>GAMMA</cite>, <cite>DELTA</cite>, <cite>HEAVY</cite>, <cite>LIGHT</cite>, <cite>KAPPA</cite>. Any column namme other than the sequence, v/j genes and identifier will be set as metadata fields to the Receptors, and can subsequently be used as labels in immuneML instructions. For TCR alpha-beta receptor import, a column mapping could for example look like this:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">cdr3_a_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">alpha_amino_acid_sequence</span>
<span class="nt">cdr3_b_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta_amino_acid_sequence</span>
<span class="nt">cdr3_a_nucseq</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">alpha_nucleotide_sequence</span>
<span class="nt">cdr3_b_nucseq</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta_nucleotide_sequence</span>
<span class="nt">v_a_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">alpha_v_call</span>
<span class="nt">v_b_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta_v_call</span>
<span class="nt">j_a_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">alpha_j_call</span>
<span class="nt">j_b_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta_j_call</span>
<span class="nt">clone_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">identifier</span>
<span class="nt">count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="nt">epitope</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope</span><span class="w"> </span><span class="c1"># metadata field</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded.</p></li>
<li><p>columns_to_load (list): Optional; specifies which columns to load from the input file. This may be useful if the input files contain many unused columns. If no value is specified, all columns are loaded.</p></li>
<li><p>separator (str): Required parameter. Column separator, for example “t” or “,”.</p></li>
<li><p>organism (str): The organism that the receptors came from. This will be set as a parameter in the ReceptorDataset object.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_receptor_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">SingleLineReceptor</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">receptor_chains</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRA_TRB</span><span class="w"> </span><span class="c1"># what chain pair to import</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># filter out sequences if they don&#39;t have sequence_aa set</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">columns_to_load</span><span class="p">:</span><span class="w"> </span><span class="c1"># which subset of columns to load from the file</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">subject</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">count</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_a_gene</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_a_gene</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cdr3_a_aa</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_b_gene</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_b_gene</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cdr3_b_aa</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">clone_id</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping file: immuneML</span>
<span class="w">            </span><span class="nt">cdr3_a_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">alpha_amino_acid_sequence</span>
<span class="w">            </span><span class="nt">cdr3_b_aa</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta_amino_acid_sequence</span>
<span class="w">            </span><span class="nt">cdr3_a_nucseq</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">alpha_nucleotide_sequence</span>
<span class="w">            </span><span class="nt">cdr3_b_nucseq</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta_nucleotide_sequence</span>
<span class="w">            </span><span class="nt">v_a_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">alpha_v_call</span>
<span class="w">            </span><span class="nt">v_b_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta_v_call</span>
<span class="w">            </span><span class="nt">j_a_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">alpha_j_call</span>
<span class="w">            </span><span class="nt">j_b_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">beta_j_call</span>
<span class="w">            </span><span class="nt">count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="w">            </span><span class="nt">clone_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">identifier</span>
<span class="w">            </span><span class="nt">epitope</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope</span>
<span class="w">            </span><span class="nt">organism</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">mouse</span>
</pre></div>
</div>
</section>
<section id="tenxgenomics">
<h4>TenxGenomics<a class="headerlink" href="#tenxgenomics" title="Link to this heading">¶</a></h4>
<p>Imports data from the 10xGenomics Cell Ranger analysis pipeline into a Repertoire-, Sequence- or ReceptorDataset.
RepertoireDatasets should be used when making predictions per repertoire, such as predicting a disease state.
SequenceDatasets or ReceptorDatasets should be used when predicting values for unpaired (single-chain) and paired
immune receptors respectively, like antigen specificity.</p>
<p>The files that should be used as input are named ‘Clonotype consensus annotations (CSV)’,
as described here: <a class="reference external" href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation#consensus">https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation#consensus</a></p>
<p>Note: by default the 10xGenomics field ‘umis’ is used to define the immuneML field counts. If you want to use the 10x Genomics
field reads instead, this can be changed in the column_mapping (set reads: counts).
Furthermore, the 10xGenomics field clonotype_id is used for the immuneML field cell_id.</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): For RepertoireDatasets, this is the path to a directory with 10xGenomics files to import. For Sequence- or ReceptorDatasets this path may either be the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset or ReceptorDataset. By default, is_repertoire is set to True.</p></li>
<li><p>metadata_file (str): Required for RepertoireDatasets. This parameter specifies the path to the metadata file. This is a csv file with columns filename, subject_id and arbitrary other columns which can be used as labels in instructions. For setting Sequence- or ReceptorDataset metadata, metadata_file is ignored, see metadata_column_mapping instead.</p></li>
<li><p>paired (str): Required for Sequence- or ReceptorDatasets. This parameter determines whether to import a SequenceDataset (paired = False) or a ReceptorDataset (paired = True). In a ReceptorDataset, two sequences with chain types specified by receptor_chains are paired together based on the identifier given in the 10xGenomics column named ‘clonotype_id’.</p></li>
<li><p>receptor_chains (str): Required for ReceptorDatasets. Determines which pair of chains to import for each Receptor.  Valid values are <cite>TRA_TRB</cite>, <cite>TRG_TRD</cite>, <cite>IGH_IGL</cite>, <cite>IGH_IGK</cite>. If receptor_chains is not provided, the chain pair is automatically detected (only one chain pair type allowed per repertoire).</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as 10xGenomics uses IMGT junction as CDR3. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): A mapping from 10xGenomics column names to immuneML’s internal data representation. A custom column mapping can be specified here if necessary (for example; adding additional data fields if they are present in the 10xGenomics file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’].. For 10xGenomics, this is by default set to:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">cdr3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="nt">cdr3_nt</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="nt">v_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="nt">j_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="nt">umis</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="nt">clonotype_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cell_id</span>
<span class="nt">consensus_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_id</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For 10xGenomics format, there is no default column_mapping_synonyms.</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are 10xGenomics column names and values are the names that are internally used in immuneML as metadata fields. These metadata fields can be used as prediction labels for Sequence- and ReceptorDatasets. This parameter can also be used to specify sequence-level metadata columns for RepertoireDatasets, which can be used by reports. To set prediction label metadata for RepertoireDatasets, see metadata_file instead. For 10xGenomics format, there is no default metadata_column_mapping.</p></li>
<li><p>separator (str): Column separator, for 10xGenomics this is by default “,”.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_10x_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10xGenomics</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span><span class="w"> </span><span class="c1"># metadata file for RepertoireDataset</span>
<span class="w">        </span><span class="nt">paired</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to import SequenceDataset (False) or ReceptorDataset (True) when is_repertoire = False</span>
<span class="w">        </span><span class="nt">receptor_chains</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRA_TRB</span><span class="w"> </span><span class="c1"># what chain pair to import for a ReceptorDataset</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># metadata column mapping 10xGenomics: immuneML for SequenceDataset</span>
<span class="w">            </span><span class="nt">tenx_column_name1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label1</span>
<span class="w">            </span><span class="nt">tenx_column_name2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">metadata_label2</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># filter out sequences if they don&#39;t have sequence_aa set</span>
<span class="w">        </span><span class="c1"># Optional fields with 10xGenomics-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;,&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping 10xGenomics: immuneML</span>
<span class="w">            </span><span class="nt">cdr3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="w">            </span><span class="nt">cdr3_nt</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence</span>
<span class="w">            </span><span class="nt">v_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="w">            </span><span class="nt">j_gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="w">            </span><span class="nt">umis</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">duplicate_count</span>
<span class="w">            </span><span class="nt">clonotype_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cell_id</span>
<span class="w">            </span><span class="nt">consensus_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_id</span>
</pre></div>
</div>
</section>
<section id="vdjdb">
<h4>VDJdb<a class="headerlink" href="#vdjdb" title="Link to this heading">¶</a></h4>
<p>Imports data in VDJdb format into a Repertoire-, Sequence- or ReceptorDataset.
RepertoireDatasets should be used when making predictions per repertoire, such as predicting a disease state.
SequenceDatasets or ReceptorDatasets should be used when predicting values for unpaired (single-chain) and paired
immune receptors respectively, like antigen specificity.</p>
<p>Arguments:</p>
<ul>
<li><p>path (str): For RepertoireDatasets, this is the path to a directory with VDJdb files to import. For Sequence- or ReceptorDatasets this path may either be the path to the file to import, or the path to the folder locating one or multiple files with .tsv, .csv or .txt extensions. By default path is set to the current working directory.</p></li>
<li><p>is_repertoire (bool): If True, this imports a RepertoireDataset. If False, it imports a SequenceDataset or ReceptorDataset. By default, is_repertoire is set to True.</p></li>
<li><p>metadata_file (str): Required for RepertoireDatasets. This parameter specifies the path to the metadata file. This is a csv file with columns filename, subject_id and arbitrary other columns which can be used as labels in instructions. For setting Sequence- or ReceptorDataset metadata, metadata_file is ignored, see metadata_column_mapping instead.</p></li>
<li><p>paired (str): Required for Sequence- or ReceptorDatasets. This parameter determines whether to import a SequenceDataset (paired = False) or a ReceptorDataset (paired = True). In a ReceptorDataset, two sequences with chain types specified by receptor_chains are paired together based on the identifier given in the VDJdb column named ‘complex.id’.</p></li>
<li><p>receptor_chains (str): Required for ReceptorDatasets. Determines which pair of chains to import for each Receptor. Valid values are <cite>TRA_TRB</cite>, <cite>TRG_TRD</cite>, <cite>IGH_IGL</cite>, <cite>IGH_IGK</cite>. If receptor_chains is not provided, the chain pair is automatically detected (only one chain pair type allowed per repertoire).</p></li>
<li><p>import_illegal_characters (bool): Whether to import sequences that contain illegal characters, i.e., characters that do not appear in the sequence alphabet (amino acids including stop codon ‘*’, or nucleotides). When set to false, filtering is only applied to the sequence type of interest (when running immuneML in amino acid mode, only entries with illegal characters in the amino acid sequence are removed). By default import_illegal_characters is False.</p></li>
<li><p>import_empty_nt_sequences (bool): imports sequences which have an empty nucleotide sequence field; can be True or False. By default, import_empty_nt_sequences is set to True.</p></li>
<li><p>import_empty_aa_sequences (bool): imports sequences which have an empty amino acid sequence field; can be True or False; for analysis on amino acid sequences, this parameter should be False (import only non-empty amino acid sequences). By default, import_empty_aa_sequences is set to False.</p></li>
<li><p>region_type (str): Which part of the sequence to import. By default, this value is set to IMGT_CDR3. This means the first and last amino acids are removed from the CDR3 sequence, as VDJdb uses IMGT junction as CDR3. Specifying any other value will result in importing the sequences as they are. Valid values are <cite>IMGT_CDR1</cite>, <cite>IMGT_CDR2</cite>, <cite>IMGT_CDR3</cite>, <cite>IMGT_FR1</cite>, <cite>IMGT_FR2</cite>, <cite>IMGT_FR3</cite>, <cite>IMGT_FR4</cite>, <cite>IMGT_JUNCTION</cite>, <cite>FULL_SEQUENCE</cite>.</p></li>
<li><p>column_mapping (dict): A mapping from VDJdb column names to immuneML’s internal data representation. A custom column mapping can be specified here if necessary (for example; adding additional data fields if they are present in the VDJdb file, or using alternative column names). Valid immuneML fields that can be specified here are [‘sequence_aa’, ‘sequence’, ‘v_call’, ‘j_call’, ‘chain’, ‘duplicate_count’, ‘frame_type’, ‘sequence_id’, ‘cell_id’].. For VDJdb, this is by default set to:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">V</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="nt">J</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="nt">CDR3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="nt">complex.id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_id</span>
<span class="nt">Gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chain</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>column_mapping_synonyms (dict): This is a column mapping that can be used if a column could have alternative names. The formatting is the same as column_mapping. If some columns specified in column_mapping are not found in the file, the columns specified in column_mapping_synonyms are instead attempted to be loaded. For VDJdb format, there is no default column_mapping_synonyms.</p></li>
<li><p>metadata_column_mapping (dict): Specifies metadata for Sequence- and ReceptorDatasets. This should specify a mapping similar to column_mapping where keys are VDJdb column names and values are the names that are internally used in immuneML as metadata fields. This means that epitope, epitope_gene and epitope_species can be used as prediction labels for Sequence- and ReceptorDatasets. This parameter can also be used to specify sequence-level metadata columns for RepertoireDatasets, which can be used by reports. To set prediction label metadata for RepertoireDatasets, see metadata_file instead. For VDJdb format, this parameter is by default set to:</p>
<blockquote>
<div><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">Epitope</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope</span>
<span class="nt">Epitope gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope_gene</span>
<span class="nt">Epitope species</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope_species</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>separator (str): Column separator, for VDJdb this is by default “t”.</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_vdjdb_dataset</span><span class="p">:</span>
<span class="w">    </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">VDJdb</span>
<span class="w">    </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/files/</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># whether to import a RepertoireDataset</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">path/to/metadata.csv</span><span class="w"> </span><span class="c1"># metadata file for RepertoireDataset</span>
<span class="w">        </span><span class="nt">paired</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to import SequenceDataset (False) or ReceptorDataset (True) when is_repertoire = False</span>
<span class="w">        </span><span class="nt">receptor_chains</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRA_TRB</span><span class="w"> </span><span class="c1"># what chain pair to import for a ReceptorDataset</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># remove sequences with illegal characters for the sequence_type being used</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># keep sequences even though the nucleotide sequence might be empty</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># filter out sequences if they don&#39;t have sequence_aa set</span>
<span class="w">        </span><span class="c1"># Optional fields with VDJdb-specific defaults, only change when different behavior is required:</span>
<span class="w">        </span><span class="nt">separator</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="w"> </span><span class="c1"># column separator</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span><span class="w"> </span><span class="c1"># what part of the sequence to import</span>
<span class="w">        </span><span class="nt">column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># column mapping VDJdb: immuneML</span>
<span class="w">            </span><span class="nt">V</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">v_call</span>
<span class="w">            </span><span class="nt">J</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">j_call</span>
<span class="w">            </span><span class="nt">CDR3</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_aa</span>
<span class="w">            </span><span class="nt">complex.id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sequence_id</span>
<span class="w">            </span><span class="nt">Gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">chain</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span><span class="w"> </span><span class="c1"># metadata column mapping VDJdb: immuneML</span>
<span class="w">            </span><span class="nt">Epitope</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope</span>
<span class="w">            </span><span class="nt">Epitope gene</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope_gene</span>
<span class="w">            </span><span class="nt">Epitope species</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">epitope_species</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="instructions">
<h2>Instructions<a class="headerlink" href="#instructions" title="Link to this heading">¶</a></h2>
<section id="feasibilitysummary">
<h3>FeasibilitySummary<a class="headerlink" href="#feasibilitysummary" title="Link to this heading">¶</a></h3>
<p>FeasibilitySummaryInstruction instruction creates a small synthetic dataset and reports summary metrics to show if the simulation with the given
parameters is feasible. The input parameters to this analysis are the name of the simulation
(the same that can be used with LigoSim instruction later if feasibility analysis looks acceptable), and the number of sequences to
simulate for estimating the feasibility.</p>
<p>The feasibility analysis is performed for each generative model separately as these could differ in the analyses that will be reported.</p>
<p>Arguments:</p>
<ul class="simple">
<li><p>simulation (str): a name of a simulation object containing a list of SimConfigItem as specified under definitions key; defines how to combine signals with simulated data; specified under definitions</p></li>
<li><p>sequence_count (int): how many sequences to generate to estimate feasibility (default value: 100 000)</p></li>
<li><p>number_of_processes (int): for the parts of the analysis that are possible to parallelize, how many processes to use</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_feasibility_summary</span><span class="p">:</span><span class="w"> </span><span class="c1"># user-defined name of the instruction</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">FeasibilitySummary</span><span class="w"> </span><span class="c1"># which instruction to execute</span>
<span class="w">    </span><span class="nt">simulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sim1</span>
<span class="w">    </span><span class="nt">sequence_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10000</span>
</pre></div>
</div>
</section>
<section id="ligosim">
<h3>LigoSim<a class="headerlink" href="#ligosim" title="Link to this heading">¶</a></h3>
<p>LIgO simulation instruction creates a synthetic dataset from scratch based on the generative model and a set of signals provided by
the user.</p>
<p>Arguments:</p>
<ul class="simple">
<li><p>simulation (str): a name of a simulation object containing a list of SimConfigItem as specified under definitions key; defines how to combine signals with simulated data; specified under definitions</p></li>
<li><p>sequence_batch_size (int): how many sequences to generate at once using the generative model before checking for signals and filtering</p></li>
<li><p>max_iterations (int): how many iterations are allowed when creating sequences</p></li>
<li><p>export_p_gens (bool): whether to compute generation probabilities (if supported by the generative model) for sequences and include them as part of output</p></li>
<li><p>number_of_processes (int): determines how many simulation items can be simulated in parallel</p></li>
</ul>
<p>YAML specification:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">my_simulation_instruction</span><span class="p">:</span><span class="w"> </span><span class="c1"># user-defined name of the instruction</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">LIgOSim</span><span class="w"> </span><span class="c1"># which instruction to execute</span>
<span class="w">    </span><span class="nt">simulation</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sim1</span>
<span class="w">    </span><span class="nt">sequence_batch_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>
<span class="w">    </span><span class="nt">max_iterations</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>
<span class="w">    </span><span class="nt">export_p_gens</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="w">    </span><span class="nt">number_of_processes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">4</span>
</pre></div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="usecases/usecase2.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Manuscript use case 2: Limitations of conventional encoding schemes for repertoire-level binary classification when immune signals co-occur within the same AIR</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">YAML specification</a><ul>
<li><a class="reference internal" href="#definitions">Definitions</a><ul>
<li><a class="reference internal" href="#simulation">Simulation</a><ul>
<li><a class="reference internal" href="#seedmotif">SeedMotif</a></li>
<li><a class="reference internal" href="#pwm">PWM</a></li>
<li><a class="reference internal" href="#signal">Signal</a></li>
<li><a class="reference internal" href="#simulation-config">Simulation config</a></li>
<li><a class="reference internal" href="#simulation-config-item">Simulation config item</a></li>
</ul>
</li>
<li><a class="reference internal" href="#generative-models">Generative models</a><ul>
<li><a class="reference internal" href="#experimentalimport">ExperimentalImport</a></li>
<li><a class="reference internal" href="#olga">OLGA</a></li>
</ul>
</li>
<li><a class="reference internal" href="#supported-dataset-formats">Supported dataset formats</a><ul>
<li><a class="reference internal" href="#airr">AIRR</a></li>
<li><a class="reference internal" href="#generic">Generic</a></li>
<li><a class="reference internal" href="#igor">IGoR</a></li>
<li><a class="reference internal" href="#ireceptor">IReceptor</a></li>
<li><a class="reference internal" href="#immuneml">ImmuneML</a></li>
<li><a class="reference internal" href="#immunoseqrearrangement">ImmunoSEQRearrangement</a></li>
<li><a class="reference internal" href="#immunoseqsample">ImmunoSEQSample</a></li>
<li><a class="reference internal" href="#mixcr">MiXCR</a></li>
<li><a class="reference internal" href="#id3">OLGA</a></li>
<li><a class="reference internal" href="#singlelinereceptor">SingleLineReceptor</a></li>
<li><a class="reference internal" href="#tenxgenomics">TenxGenomics</a></li>
<li><a class="reference internal" href="#vdjdb">VDJdb</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#instructions">Instructions</a><ul>
<li><a class="reference internal" href="#feasibilitysummary">FeasibilitySummary</a></li>
<li><a class="reference internal" href="#ligosim">LigoSim</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=9306f2a9"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>